<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="dominoSignal">
<title>Interacting with domino Objects • dominoSignal</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Fira_Code-0.4.8/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<script src="../extra.js"></script><meta property="og:title" content="Interacting with domino Objects">
<meta property="og:description" content="dominoSignal">
<meta property="og:image" content="https://FertigLab.github.io/dominoSignal/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-#7DDED2"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">dominoSignal</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/dominoSignal.html">Get started</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/cellphonedb_vignette.html">Using the CellPhoneDB Database</a>
    <a class="dropdown-item" href="../articles/tf_scenic_vignette.html">SCENIC for TF Activation Scoring</a>
    <a class="dropdown-item" href="../articles/domino_object_vignette.html">Interacting with domino Objects</a>
    <a class="dropdown-item" href="../articles/plotting_vignette.html">Plotting Functions and Options</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-news">News</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-news">
    <h6 class="dropdown-header" data-toc-skip>Releases</h6>
    <a class="external-link dropdown-item" href="https://github.com/FertigLab/dominoSignal/releases/tag/v0.99.2-alpha">dominoSignal v0.99.2</a>
    <a class="external-link dropdown-item" href="https://github.com/FertigLab/dominoSignal/releases/tag/v0.2.1-alpha">dominoSignal v0.2.1</a>
    <a class="external-link dropdown-item" href="https://github.com/FertigLab/dominoSignal/releases/tag/v0.2.2-alpha">dominoSignal v0.2.2</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../news/index.html">Changelog</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/FertigLab/dominoSignal/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="domino_object_vignette_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Interacting with domino Objects</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/FertigLab/dominoSignal/blob/HEAD/vignettes/domino_object_vignette.Rmd" class="external-link"><code>vignettes/domino_object_vignette.Rmd</code></a></small>
      <div class="d-none name"><code>domino_object_vignette.Rmd</code></div>
    </div>

    
    
<p>In this tutorial, we will go into more detail on the structure of a domino object and the ways in which to access the data stored within. We will be using the domino object we built on the <a href="https://fertiglab.github.io/dominoSignal" class="external-link">Getting Started</a> page. If you have not yet built a domino object, you can do so by following the instructions on the <a href="https://fertiglab.github.io/dominoSignal/articles/dominoSignal" class="external-link">Getting Started</a> page.</p>
<details><p><summary>Instructions to load data</summary></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://fertiglab.github.io/dominoSignal/" class="external-link">dominoSignal</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># BiocFileCache helps with file management across sessions</span></span>
<span><span class="va">bfc</span> <span class="op">&lt;-</span> <span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">BiocFileCache</a></span><span class="op">(</span>ask <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">data_url</span> <span class="op">&lt;-</span> <span class="st">"https://zenodo.org/records/10951634/files/pbmc_domino_built.rds"</span></span>
<span><span class="va">tmp_path</span> <span class="op">&lt;-</span> <span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcrpath</a></span><span class="op">(</span><span class="va">bfc</span>, <span class="va">data_url</span><span class="op">)</span></span>
<span><span class="va">dom</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">tmp_path</span><span class="op">)</span></span></code></pre></div>
</details><div class="section level2">
<h2 id="object-contents">Object contents<a class="anchor" aria-label="anchor" href="#object-contents"></a>
</h2>
<p>There is a great deal of information stored with the <a href="https://fertiglab.github.io/dominoSignal/reference/domino-class.html" class="external-link">domino object class</a>. The domino object is an S4 class object that contains a variety of information about the data set used to build the object, the calculated values, and the linkages between receptors, ligands, and transcription factors. The object is structured as follows (with some examples of the information stored within each slot:</p>
<ul>
<li>
<p>Input Data</p>
<ul>
<li><p>Information about the database used to construct the rl_map</p></li>
<li><p>Inputted counts matrix</p></li>
<li><p>Inputted z-scored counts matrix</p></li>
<li><p>Inputted cluster labels</p></li>
<li><p>Inputted transcription factor activation scores</p></li>
</ul>
</li>
<li>
<p>Calculated values</p>
<ul>
<li><p>Differential expression p-values of transcription factors in each cluster</p></li>
<li><p>Correlation values between ligands and receptors</p></li>
<li><p>Median correlation between components of receptor complexes</p></li>
</ul>
</li>
<li>
<p>Linkages</p>
<ul>
<li><p>Complexes show the component genes of any complexes in the rl map</p></li>
<li><p>Receptor - ligand linkages as determined from the rl map</p></li>
<li><p>Transcription factor - target linkages as determined from the SCENIC analysis (or other regulon inference method)</p></li>
<li><p>Transcription factors that are differentially expressed in each cluster</p></li>
<li><p>Transcription factors that are correlated with receptors</p></li>
<li><p>Transcription factors that are correlated with receptors in each cluster</p></li>
<li><p>Receptors which are active in each cluster</p></li>
<li><p>Ligands that may activate a receptor in a given cluster (so-called incoming ligands; these may include ligands from outside the data set)</p></li>
</ul>
</li>
<li>
<p>Signaling matrices</p>
<ul>
<li><p>For each cluster, incoming ligands and the clusters within the data set that they are coming from</p></li>
<li><p>A summary of signaling between all clusters</p></li>
</ul>
</li>
<li>
<p>Miscellaneous Information</p>
<ul>
<li><p>Build information, which includes the parameters used to build the object in the <code><a href="../reference/build_domino.html">build_domino()</a></code> functions</p></li>
<li><p>The pared down receptor ligand map information used in building the object</p></li>
<li><p>The percent expression of receptors within each cluster</p></li>
</ul>
</li>
</ul>
<p>For commonly accessed information (the number of cells, clusters, and some build information), the show and print methods for domino objects can be used.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dom</span></span>
<span><span class="co">#&gt; A domino object of 2607 cells</span></span>
<span><span class="co">#&gt; Built with signaling between 9 clusters</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">dom</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="access-functions">Access functions<a class="anchor" aria-label="anchor" href="#access-functions"></a>
</h2>
<p>To facilitate access to the information stored in the domino object, we have provided a collection of functions to retrieve specific items. These functions begin with “dom_” and can be listed using <code><a href="https://rdrr.io/r/base/ls.html" class="external-link">ls()</a></code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/ls.html" class="external-link">ls</a></span><span class="op">(</span><span class="st">"package:dominoSignal"</span>, pattern <span class="op">=</span> <span class="st">"^dom_"</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "dom_clusters"      "dom_correlations"  "dom_counts"       </span></span>
<span><span class="co">#&gt;  [4] "dom_database"      "dom_de"            "dom_info"         </span></span>
<span><span class="co">#&gt;  [7] "dom_linkages"      "dom_network_items" "dom_signaling"    </span></span>
<span><span class="co">#&gt; [10] "dom_tf_activation" "dom_zscores"</span></span></code></pre></div>
<div class="section level3">
<h3 id="input-data">Input data<a class="anchor" aria-label="anchor" href="#input-data"></a>
</h3>
<p>When creating a domino object with the <code><a href="../reference/create_domino.html">create_domino()</a></code> function, several inputs are required which are then stored in the domino object itself. These include cluster labels, the counts matrix, z-scored counts matrix, transcription factor activation scores, and the R-L database used in <code><a href="../reference/create_rl_map_cellphonedb.html">create_rl_map_cellphonedb()</a></code>.</p>
<p>For example, to access the cluster names in the domino object:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/dom_clusters.html">dom_clusters</a></span><span class="op">(</span><span class="va">dom</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "B_cell"            "CD14_monocyte"     "CD16_monocyte"    </span></span>
<span><span class="co">#&gt; [4] "CD8_T_cell"        "dendritic_cell"    "memory_CD4_T_cell"</span></span>
<span><span class="co">#&gt; [7] "naive_CD4_T_cell"  "NK_cell"           "Platelet"</span></span></code></pre></div>
<p>Setting an argument <code>labels = TRUE</code> will return the vector of cluster labels for each cell rather than the unique cluster names.</p>
<p>To access the counts:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">count_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dom_counts.html">dom_counts</a></span><span class="op">(</span><span class="va">dom</span><span class="op">)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">count_matrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="14%">
<col width="17%">
<col width="17%">
<col width="17%">
<col width="17%">
<col width="17%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="right">AAACATACAACCAC-1</th>
<th align="right">AAACATTGAGCTAC-1</th>
<th align="right">AAACATTGATCAGC-1</th>
<th align="right">AAACCGTGCTTCCG-1</th>
<th align="right">AAACCGTGTATGCG-1</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">AL627309.1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">AP006222.2</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">RP11-206L10.2</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">RP11-206L10.9</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">FAM87B</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
<p>Or z-scored counts:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">z_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dom_zscores.html">dom_zscores</a></span><span class="op">(</span><span class="va">dom</span><span class="op">)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">z_matrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="14%">
<col width="17%">
<col width="17%">
<col width="17%">
<col width="17%">
<col width="17%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="right">AAACATACAACCAC-1</th>
<th align="right">AAACATTGAGCTAC-1</th>
<th align="right">AAACATTGATCAGC-1</th>
<th align="right">AAACCGTGCTTCCG-1</th>
<th align="right">AAACCGTGTATGCG-1</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">AL627309.1</td>
<td align="right">-0.0581122</td>
<td align="right">-0.0581122</td>
<td align="right">-0.0581122</td>
<td align="right">-0.0581122</td>
<td align="right">-0.0581122</td>
</tr>
<tr class="even">
<td align="left">AP006222.2</td>
<td align="right">-0.0335151</td>
<td align="right">-0.0335151</td>
<td align="right">-0.0335151</td>
<td align="right">-0.0335151</td>
<td align="right">-0.0335151</td>
</tr>
<tr class="odd">
<td align="left">RP11-206L10.2</td>
<td align="right">-0.0399375</td>
<td align="right">-0.0399375</td>
<td align="right">-0.0399375</td>
<td align="right">-0.0399375</td>
<td align="right">-0.0399375</td>
</tr>
<tr class="even">
<td align="left">RP11-206L10.9</td>
<td align="right">-0.0337574</td>
<td align="right">-0.0337574</td>
<td align="right">-0.0337574</td>
<td align="right">-0.0337574</td>
<td align="right">-0.0337574</td>
</tr>
<tr class="odd">
<td align="left">FAM87B</td>
<td align="right">-0.0274227</td>
<td align="right">-0.0274227</td>
<td align="right">-0.0274227</td>
<td align="right">-0.0274227</td>
<td align="right">-0.0274227</td>
</tr>
</tbody>
</table>
<p>The transcription factor activation scores can be similarly accessed:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">activation_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dom_tf_activation.html">dom_tf_activation</a></span><span class="op">(</span><span class="va">dom</span><span class="op">)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">activation_matrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="6%">
<col width="18%">
<col width="18%">
<col width="18%">
<col width="18%">
<col width="18%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="right">AAACATACAACCAC-1</th>
<th align="right">AAACATTGAGCTAC-1</th>
<th align="right">AAACATTGATCAGC-1</th>
<th align="right">AAACCGTGCTTCCG-1</th>
<th align="right">AAACCGTGTATGCG-1</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">ARNTL</td>
<td align="right">0.0446386</td>
<td align="right">0.0413253</td>
<td align="right">0.0441566</td>
<td align="right">0.0437952</td>
<td align="right">0.1087952</td>
</tr>
<tr class="even">
<td align="left">ATF3</td>
<td align="right">0.0582029</td>
<td align="right">0.0870623</td>
<td align="right">0.0509289</td>
<td align="right">0.1172953</td>
<td align="right">0.0638104</td>
</tr>
<tr class="odd">
<td align="left">ATF4</td>
<td align="right">0.0295733</td>
<td align="right">0.0437047</td>
<td align="right">0.0542889</td>
<td align="right">0.0427879</td>
<td align="right">0.0436855</td>
</tr>
<tr class="even">
<td align="left">ATF6</td>
<td align="right">0.0000000</td>
<td align="right">0.0211198</td>
<td align="right">0.1151665</td>
<td align="right">0.0850461</td>
<td align="right">0.0460666</td>
</tr>
<tr class="odd">
<td align="left">BCL3</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
</tr>
</tbody>
</table>
<p>Information about the database referenced for ligand - receptor pairs and composition of protein complexes can be extracted from the <code><a href="../reference/dom_database.html">dom_database()</a></code> function. By default, the function returns the name(s) of the database(s) used:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/dom_database.html">dom_database</a></span><span class="op">(</span><span class="va">dom</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "CellPhoneDB_v4.0"</span></span></code></pre></div>
<p>If you would like to view the entire ligand - receptor map, set <code>name_only = FALSE</code>:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">db_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dom_database.html">dom_database</a></span><span class="op">(</span><span class="va">dom</span>, name_only <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">db_matrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="3%">
<col width="45%">
<col width="16%">
<col width="14%">
<col width="12%">
<col width="7%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="left">int_pair</th>
<th align="left">name_A</th>
<th align="left">uniprot_A</th>
<th align="left">gene_A</th>
<th align="left">type_A</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">4</td>
<td align="left">RAreceptor_RXRG &amp; atRetinoicAcid_byALDH1A3</td>
<td align="left">RAreceptor_RXRG</td>
<td align="left">P48443,P29373</td>
<td align="left">RXRG,CRABP2</td>
<td align="left">R</td>
</tr>
<tr class="even">
<td align="left">5</td>
<td align="left">RAreceptor_RXRG &amp; atRetinoicAcid_byALDH1A2</td>
<td align="left">RAreceptor_RXRG</td>
<td align="left">P48443,P29373</td>
<td align="left">RXRG,CRABP2</td>
<td align="left">R</td>
</tr>
<tr class="odd">
<td align="left">6</td>
<td align="left">RAreceptor_RXRG &amp; atRetinoicAcid_byALDH1A1</td>
<td align="left">RAreceptor_RXRG</td>
<td align="left">P48443,P29373</td>
<td align="left">RXRG,CRABP2</td>
<td align="left">R</td>
</tr>
<tr class="even">
<td align="left">7</td>
<td align="left">RAreceptor_RXRB &amp; atRetinoicAcid_byALDH1A3</td>
<td align="left">RAreceptor_RXRB</td>
<td align="left">P28702,P29373</td>
<td align="left">RXRB,CRABP2</td>
<td align="left">R</td>
</tr>
<tr class="odd">
<td align="left">8</td>
<td align="left">RAreceptor_RXRB &amp; atRetinoicAcid_byALDH1A2</td>
<td align="left">RAreceptor_RXRB</td>
<td align="left">P28702,P29373</td>
<td align="left">RXRB,CRABP2</td>
<td align="left">R</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="calculations">Calculations<a class="anchor" aria-label="anchor" href="#calculations"></a>
</h3>
<p>Active transcription factors in each cluster are determined by conducting Wilcoxon rank sum tests for each transcription factor where the transcription factor activity scores amongst all cells in a cluster are tested against the activity scores of all cells outside of the cluster. The p-values for the one-sided test for greater activity within the cluster compared to other cells can be accessed with the <code><a href="../reference/dom_de.html">dom_de()</a></code> function.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">de_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dom_de.html">dom_de</a></span><span class="op">(</span><span class="va">dom</span><span class="op">)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">de_matrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">B_cell</th>
<th align="right">CD14_monocyte</th>
<th align="right">CD16_monocyte</th>
<th align="right">CD8_T_cell</th>
<th align="right">dendritic_cell</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">ARNTL</td>
<td align="right">0.0000019</td>
<td align="right">1</td>
<td align="right">0.9947780</td>
<td align="right">0.9865518</td>
<td align="right">0.8359381</td>
</tr>
<tr class="even">
<td align="left">ATF3</td>
<td align="right">0.0006972</td>
<td align="right">0</td>
<td align="right">0.0000000</td>
<td align="right">1.0000000</td>
<td align="right">0.0000000</td>
</tr>
<tr class="odd">
<td align="left">ATF4</td>
<td align="right">0.8399818</td>
<td align="right">1</td>
<td align="right">1.0000000</td>
<td align="right">0.9631934</td>
<td align="right">0.1748338</td>
</tr>
<tr class="even">
<td align="left">ATF6</td>
<td align="right">0.5988625</td>
<td align="right">1</td>
<td align="right">0.9893142</td>
<td align="right">0.0365779</td>
<td align="right">0.9549489</td>
</tr>
<tr class="odd">
<td align="left">BCL3</td>
<td align="right">0.9299696</td>
<td align="right">0</td>
<td align="right">0.6007326</td>
<td align="right">0.1482247</td>
<td align="right">0.0757460</td>
</tr>
</tbody>
</table>
<p>Linkage between receptors and transcription factors is assessed by Spearman correlation between transcription factor activity scores and scaled expression of receptor-encoding genes across all cells in the data set. Spearman coefficients can be accessed with the <code><a href="../reference/dom_correlations.html">dom_correlations()</a></code> function. Setting <code>type</code> to “complex” will return the median correlation between components of receptor complexes; the default (“rl”) will return receptor - ligand correlations.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cor_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dom_correlations.html">dom_correlations</a></span><span class="op">(</span><span class="va">dom</span><span class="op">)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">cor_matrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">ARNTL</th>
<th align="right">ATF3</th>
<th align="right">ATF4</th>
<th align="right">ATF6</th>
<th align="right">BCL3</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">TNFRSF18</td>
<td align="right">0.0057014</td>
<td align="right">-0.0297120</td>
<td align="right">0.0043825</td>
<td align="right">0.0063055</td>
<td align="right">-0.0029035</td>
</tr>
<tr class="even">
<td align="left">TNFRSF4</td>
<td align="right">0.0373481</td>
<td align="right">-0.1012705</td>
<td align="right">0.0344042</td>
<td align="right">0.0508720</td>
<td align="right">0.0282500</td>
</tr>
<tr class="odd">
<td align="left">TNFRSF14</td>
<td align="right">-0.0356892</td>
<td align="right">0.0541435</td>
<td align="right">0.0015811</td>
<td align="right">0.0093978</td>
<td align="right">0.0206700</td>
</tr>
<tr class="even">
<td align="left">TNFRSF25</td>
<td align="right">0.0310362</td>
<td align="right">-0.1171502</td>
<td align="right">0.0701541</td>
<td align="right">0.0458227</td>
<td align="right">0.0047186</td>
</tr>
<tr class="odd">
<td align="left">TNFRSF1B</td>
<td align="right">-0.0933019</td>
<td align="right">0.2690328</td>
<td align="right">-0.0625319</td>
<td align="right">-0.0457366</td>
<td align="right">0.0239629</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="linkages">Linkages<a class="anchor" aria-label="anchor" href="#linkages"></a>
</h3>
<p>Linkages between ligands, receptors, and transcription factors can be accessed in several different ways, depending on the specific link and the scope desired. The <code><a href="../reference/dom_linkages.html">dom_linkages()</a></code> function has three arguments - the first, like all of our access functions, is for the domino object. The second, <code>link_type</code>, is used to specify which linkages are desired (options are complexes, receptor - ligand, tf - target, or tf - receptor). The third argument, <code>by_cluster</code>, determines whether the linkages returned are arranged by cluster (though this does change the available linkage types to tf - receptor, receptor, or incoming-ligand). For example, to access the complexes used across the dataset:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">complex_links</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dom_linkages.html">dom_linkages</a></span><span class="op">(</span><span class="va">dom</span>, link_type <span class="op">=</span> <span class="st">"complexes"</span><span class="op">)</span></span>
<span><span class="co"># Look for components of NODAL receptor complex</span></span>
<span><span class="va">complex_links</span><span class="op">$</span><span class="va">NODAL_receptor</span></span>
<span><span class="co">#&gt; NULL</span></span></code></pre></div>
<p>To view incoming ligands to each cluster:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">incoming_links</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dom_linkages.html">dom_linkages</a></span><span class="op">(</span><span class="va">dom</span>, link_type <span class="op">=</span> <span class="st">"incoming-ligand"</span>, by_cluster <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co"># Check incoming signals to dendritic cells</span></span>
<span><span class="va">incoming_links</span><span class="op">$</span><span class="va">dendritic_cell</span></span>
<span><span class="co">#&gt;  [1] "COPA"                  "MIF"                   "APP"                  </span></span>
<span><span class="co">#&gt;  [4] "FAM19A4"               "TAFA4"                 "ANXA1"                </span></span>
<span><span class="co">#&gt;  [7] "CD99"                  "integrin_aVb3_complex" "integrin_a4b1_complex"</span></span>
<span><span class="co">#&gt; [10] "BMP8B"                 "PLAU"                  "CSF3"                 </span></span>
<span><span class="co">#&gt; [13] "CXCL9"                 "HLA-F"                 "CD1D"                 </span></span>
<span><span class="co">#&gt; [16] "INS"                   "IL34"                  "CSF1"                 </span></span>
<span><span class="co">#&gt; [19] "CSF2"                  "CTLA4"                 "CD28"                 </span></span>
<span><span class="co">#&gt; [22] "GRN"                   "TNF"                   "LTA"                  </span></span>
<span><span class="co">#&gt; [25] "CXCL12"                "CXCL14"                "CD58"</span></span></code></pre></div>
<p>If, for some reason, you find yourself in need of the entire linkage structure (not recommended), it can be accessed through its slot name; domino objects are S4 objects.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">all_linkages</span> <span class="op">&lt;-</span> <span class="fu">slot</span><span class="op">(</span><span class="va">dom</span>, <span class="st">"linkages"</span><span class="op">)</span></span>
<span><span class="co"># Names of all sub-structures:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">all_linkages</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "complexes"          "rec_lig"            "tf_targets"        </span></span>
<span><span class="co">#&gt; [4] "clust_tf"           "tf_rec"             "clust_tf_rec"      </span></span>
<span><span class="co">#&gt; [7] "clust_rec"          "clust_incoming_lig"</span></span></code></pre></div>
<p>Alternately, to obtain a simplified list of receptors, ligands, and/or features in the domino object, use the <code><a href="../reference/dom_network_items.html">dom_network_items()</a></code> function. To pull all transcription factors associated with the dendritic cell cluster:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dc_tfs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dom_network_items.html">dom_network_items</a></span><span class="op">(</span><span class="va">dom</span>, <span class="st">"dendritic_cell"</span>, return <span class="op">=</span> <span class="st">"features"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">dc_tfs</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "ATF3"  "CEBPD" "FOSB"  "STAT6" "KLF4"  "CEBPA"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="signaling-matrices">Signaling Matrices<a class="anchor" aria-label="anchor" href="#signaling-matrices"></a>
</h3>
<p>The averaged z-scored expression of ligands and receptors between different clusters can be accessed in matrix form.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">signal_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dom_signaling.html">dom_signaling</a></span><span class="op">(</span><span class="va">dom</span><span class="op">)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">signal_matrix</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="13%">
<col width="6%">
<col width="10%">
<col width="10%">
<col width="8%">
<col width="11%">
<col width="13%">
<col width="12%">
<col width="6%">
<col width="7%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="right">L_B_cell</th>
<th align="right">L_CD14_monocyte</th>
<th align="right">L_CD16_monocyte</th>
<th align="right">L_CD8_T_cell</th>
<th align="right">L_dendritic_cell</th>
<th align="right">L_memory_CD4_T_cell</th>
<th align="right">L_naive_CD4_T_cell</th>
<th align="right">L_NK_cell</th>
<th align="right">L_Platelet</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">R_B_cell</td>
<td align="right">0.0494729</td>
<td align="right">0.1976263</td>
<td align="right">0.1686996</td>
<td align="right">0.0000000</td>
<td align="right">0.3040989</td>
<td align="right">0.2459618</td>
<td align="right">0.1367747</td>
<td align="right">0.0000000</td>
<td align="right">4.013299</td>
</tr>
<tr class="even">
<td align="left">R_CD14_monocyte</td>
<td align="right">0.8060665</td>
<td align="right">2.5378490</td>
<td align="right">1.4429140</td>
<td align="right">1.1787148</td>
<td align="right">3.4961650</td>
<td align="right">2.2611486</td>
<td align="right">1.2608492</td>
<td align="right">1.4485708</td>
<td align="right">10.559449</td>
</tr>
<tr class="odd">
<td align="left">R_CD16_monocyte</td>
<td align="right">0.8060665</td>
<td align="right">2.5378490</td>
<td align="right">1.1473216</td>
<td align="right">1.1787148</td>
<td align="right">3.4961650</td>
<td align="right">2.2611486</td>
<td align="right">0.8757246</td>
<td align="right">1.4055805</td>
<td align="right">7.258572</td>
</tr>
<tr class="even">
<td align="left">R_CD8_T_cell</td>
<td align="right">0.3759111</td>
<td align="right">0.0000000</td>
<td align="right">0.1263349</td>
<td align="right">0.2120444</td>
<td align="right">0.0000000</td>
<td align="right">0.2772703</td>
<td align="right">0.0000000</td>
<td align="right">0.1604005</td>
<td align="right">0.000000</td>
</tr>
<tr class="odd">
<td align="left">R_dendritic_cell</td>
<td align="right">0.3176857</td>
<td align="right">2.5041485</td>
<td align="right">0.7368126</td>
<td align="right">1.3152485</td>
<td align="right">3.4961650</td>
<td align="right">2.3809976</td>
<td align="right">0.3061580</td>
<td align="right">1.5659811</td>
<td align="right">7.258572</td>
</tr>
<tr class="even">
<td align="left">R_memory_CD4_T_cell</td>
<td align="right">0.3759111</td>
<td align="right">0.0000000</td>
<td align="right">0.1263349</td>
<td align="right">0.2120444</td>
<td align="right">0.0000000</td>
<td align="right">0.3801282</td>
<td align="right">0.0000000</td>
<td align="right">0.1604005</td>
<td align="right">0.000000</td>
</tr>
<tr class="odd">
<td align="left">R_naive_CD4_T_cell</td>
<td align="right">0.3759111</td>
<td align="right">0.0000000</td>
<td align="right">0.1263349</td>
<td align="right">0.2120444</td>
<td align="right">0.0000000</td>
<td align="right">0.3801282</td>
<td align="right">0.0000000</td>
<td align="right">0.1604005</td>
<td align="right">0.000000</td>
</tr>
<tr class="even">
<td align="left">R_NK_cell</td>
<td align="right">0.2930600</td>
<td align="right">1.5760509</td>
<td align="right">0.6086536</td>
<td align="right">0.2556741</td>
<td align="right">1.5025648</td>
<td align="right">0.7620921</td>
<td align="right">0.1772729</td>
<td align="right">0.1604005</td>
<td align="right">4.013299</td>
</tr>
<tr class="odd">
<td align="left">R_Platelet</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.000000</td>
</tr>
</tbody>
</table>
<p>To view signaling to a specific cluster from the other clusters, set the <code>cluster</code> argument to the cluster name.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dc_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dom_signaling.html">dom_signaling</a></span><span class="op">(</span><span class="va">dom</span>, <span class="st">"dendritic_cell"</span><span class="op">)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">dc_matrix</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="7%">
<col width="8%">
<col width="10%">
<col width="10%">
<col width="9%">
<col width="11%">
<col width="12%">
<col width="11%">
<col width="8%">
<col width="9%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="right">dendritic_cell.L_B_cell</th>
<th align="right">dendritic_cell.L_CD14_monocyte</th>
<th align="right">dendritic_cell.L_CD16_monocyte</th>
<th align="right">dendritic_cell.L_CD8_T_cell</th>
<th align="right">dendritic_cell.L_dendritic_cell</th>
<th align="right">dendritic_cell.L_memory_CD4_T_cell</th>
<th align="right">dendritic_cell.L_naive_CD4_T_cell</th>
<th align="right">dendritic_cell.L_NK_cell</th>
<th align="right">dendritic_cell.L_Platelet</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">COPA</td>
<td align="right">0.0000000</td>
<td align="right">0.1352458</td>
<td align="right">0.1686996</td>
<td align="right">0.0000000</td>
<td align="right">0.1706852</td>
<td align="right">0.0243003</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.1634935</td>
</tr>
<tr class="even">
<td align="left">MIF</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0810828</td>
<td align="right">0.1637432</td>
<td align="right">0.1367747</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
</tr>
<tr class="odd">
<td align="left">APP</td>
<td align="right">0.0494729</td>
<td align="right">0.0623805</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0523309</td>
<td align="right">0.0579182</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">3.8498060</td>
</tr>
<tr class="even">
<td align="left">ANXA1</td>
<td align="right">0.0000000</td>
<td align="right">0.1864958</td>
<td align="right">0.0000000</td>
<td align="right">0.1682595</td>
<td align="right">0.6876552</td>
<td align="right">0.5065038</td>
<td align="right">0.0000000</td>
<td align="right">0.3542401</td>
<td align="right">0.0000000</td>
</tr>
<tr class="odd">
<td align="left">CD99</td>
<td align="right">0.0000000</td>
<td align="right">0.0626658</td>
<td align="right">0.0000000</td>
<td align="right">0.4718174</td>
<td align="right">0.0000000</td>
<td align="right">0.1952307</td>
<td align="right">0.0000000</td>
<td align="right">0.7924524</td>
<td align="right">1.9983051</td>
</tr>
<tr class="even">
<td align="left">integrin_a4b1_complex</td>
<td align="right">0.0775377</td>
<td align="right">0.0000000</td>
<td align="right">0.0274081</td>
<td align="right">0.0128595</td>
<td align="right">0.0933263</td>
<td align="right">0.1408531</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">1.2469674</td>
</tr>
<tr class="odd">
<td align="left">BMP8B</td>
<td align="right">0.1291568</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
</tr>
<tr class="even">
<td align="left">PLAU</td>
<td align="right">0.0000000</td>
<td align="right">0.0455798</td>
<td align="right">0.0000000</td>
<td align="right">0.0711901</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
</tr>
<tr class="odd">
<td align="left">CXCL9</td>
<td align="right">0.0000000</td>
<td align="right">0.0663249</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0582573</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
</tr>
<tr class="even">
<td align="left">HLA-F</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.2918047</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.2588880</td>
<td align="right">0.0000000</td>
</tr>
<tr class="odd">
<td align="left">CD1D</td>
<td align="right">0.0615183</td>
<td align="right">0.5670314</td>
<td align="right">0.1007509</td>
<td align="right">0.0000000</td>
<td align="right">1.2126187</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
</tr>
<tr class="even">
<td align="left">CSF1</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0067252</td>
<td align="right">0.0000000</td>
<td align="right">0.1724684</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
</tr>
<tr class="odd">
<td align="left">CTLA4</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0369179</td>
<td align="right">0.0000000</td>
<td align="right">0.2085753</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
</tr>
<tr class="even">
<td align="left">CD28</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.2771802</td>
<td align="right">0.1288850</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
</tr>
<tr class="odd">
<td align="left">GRN</td>
<td align="right">0.0000000</td>
<td align="right">1.3784246</td>
<td align="right">0.2818176</td>
<td align="right">0.0000000</td>
<td align="right">1.1878101</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
</tr>
<tr class="even">
<td align="left">TNF</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0318015</td>
<td align="right">0.1191403</td>
<td align="right">0.0106558</td>
<td align="right">0.1381279</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
</tr>
<tr class="odd">
<td align="left">LTA</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.1828448</td>
<td align="right">0.0404983</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
</tr>
<tr class="even">
<td align="left">CXCL12</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.1028579</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
</tr>
<tr class="odd">
<td align="left">CD58</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.1263349</td>
<td align="right">0.1365337</td>
<td align="right">0.0000000</td>
<td align="right">0.1521363</td>
<td align="right">0.0000000</td>
<td align="right">0.1604005</td>
<td align="right">0.0000000</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="build-information">Build information<a class="anchor" aria-label="anchor" href="#build-information"></a>
</h3>
<p>To keep track of the options set when running <code><a href="../reference/build_domino.html">build_domino()</a></code>, they are stored within the domino object itself. To view these options, use the <code><a href="../reference/dom_info.html">dom_info()</a></code> function.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/dom_info.html">dom_info</a></span><span class="op">(</span><span class="va">dom</span><span class="op">)</span></span>
<span><span class="co">#&gt; $create</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $build</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $build_variables</span></span>
<span><span class="co">#&gt;     max_tf_per_clust          min_tf_pval       max_rec_per_tf </span></span>
<span><span class="co">#&gt;               25.000                0.001               25.000 </span></span>
<span><span class="co">#&gt; rec_tf_cor_threshold   min_rec_percentage </span></span>
<span><span class="co">#&gt;                0.250                0.100</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="continued-development">Continued Development<a class="anchor" aria-label="anchor" href="#continued-development"></a>
</h2>
<p>Since dominoSignal is a package still being developed, there are new functions and features that will be implemented in future versions. In the meantime, we have put together further information on <a href="https://fertiglab.github.io/dominoSignal/articles/plotting_vignette" class="external-link">plotting</a> and an example analysis can be viewed on our <a href="https://fertiglab.github.io/dominoSignal/articles/dominoSignal" class="external-link">Getting Started</a> page. Additionally, if you find any bugs, have further questions, or want to share an idea, please let us know <a href="https://github.com/FertigLab/dominoSignal/issues" class="external-link">here</a>.</p>
<details><p><summary>Vignette Build Information</summary> Date last built and session information:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.Date</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "2024-06-25"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0</span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] dominoSignal_0.99.2</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] bitops_1.0-7           matrixStats_1.2.0      fs_1.6.3              </span></span>
<span><span class="co">#&gt;   [4] bit64_4.0.5            filelock_1.0.3         doParallel_1.0.17     </span></span>
<span><span class="co">#&gt;   [7] RColorBrewer_1.1-3     progress_1.2.3         httr_1.4.7            </span></span>
<span><span class="co">#&gt;  [10] GenomeInfoDb_1.34.9    backports_1.4.1        tools_4.2.1           </span></span>
<span><span class="co">#&gt;  [13] bslib_0.6.1            utf8_1.2.4             R6_2.5.1              </span></span>
<span><span class="co">#&gt;  [16] DBI_1.2.2              BiocGenerics_0.44.0    colorspace_2.1-0      </span></span>
<span><span class="co">#&gt;  [19] GetoptLong_1.0.5       withr_3.0.0            tidyselect_1.2.1      </span></span>
<span><span class="co">#&gt;  [22] prettyunits_1.2.0      bit_4.0.5              curl_5.2.1            </span></span>
<span><span class="co">#&gt;  [25] compiler_4.2.1         textshaping_0.3.7      cli_3.6.2             </span></span>
<span><span class="co">#&gt;  [28] Biobase_2.58.0         formatR_1.14           xml2_1.3.6            </span></span>
<span><span class="co">#&gt;  [31] desc_1.4.3             sass_0.4.8             scales_1.3.0          </span></span>
<span><span class="co">#&gt;  [34] rappdirs_0.3.3         pkgdown_2.0.7          systemfonts_1.0.6     </span></span>
<span><span class="co">#&gt;  [37] stringr_1.5.1          digest_0.6.35          rmarkdown_2.26        </span></span>
<span><span class="co">#&gt;  [40] XVector_0.38.0         pkgconfig_2.0.3        htmltools_0.5.7       </span></span>
<span><span class="co">#&gt;  [43] dbplyr_2.5.0           fastmap_1.1.1          rlang_1.1.3           </span></span>
<span><span class="co">#&gt;  [46] GlobalOptions_0.1.2    RSQLite_2.3.5          shape_1.4.6.1         </span></span>
<span><span class="co">#&gt;  [49] jquerylib_0.1.4        generics_0.1.3         jsonlite_1.8.8        </span></span>
<span><span class="co">#&gt;  [52] car_3.1-2              dplyr_1.1.4            RCurl_1.98-1.14       </span></span>
<span><span class="co">#&gt;  [55] magrittr_2.0.3         GenomeInfoDbData_1.2.9 Matrix_1.6-5          </span></span>
<span><span class="co">#&gt;  [58] munsell_0.5.0          Rcpp_1.0.12            S4Vectors_0.36.2      </span></span>
<span><span class="co">#&gt;  [61] fansi_1.0.6            abind_1.4-5            lifecycle_1.0.4       </span></span>
<span><span class="co">#&gt;  [64] stringi_1.8.3          yaml_2.3.8             carData_3.0-5         </span></span>
<span><span class="co">#&gt;  [67] zlibbioc_1.44.0        plyr_1.8.9             BiocFileCache_2.11.2  </span></span>
<span><span class="co">#&gt;  [70] grid_4.2.1             blob_1.2.4             parallel_4.2.1        </span></span>
<span><span class="co">#&gt;  [73] crayon_1.5.2           lattice_0.22-5         Biostrings_2.66.0     </span></span>
<span><span class="co">#&gt;  [76] circlize_0.4.16        hms_1.1.3              KEGGREST_1.38.0       </span></span>
<span><span class="co">#&gt;  [79] knitr_1.45             ComplexHeatmap_2.14.0  pillar_1.9.0          </span></span>
<span><span class="co">#&gt;  [82] igraph_2.0.3           ggpubr_0.6.0           rjson_0.2.21          </span></span>
<span><span class="co">#&gt;  [85] ggsignif_0.6.4         codetools_0.2-19       biomaRt_2.54.1        </span></span>
<span><span class="co">#&gt;  [88] stats4_4.2.1           XML_3.99-0.16.1        glue_1.7.0            </span></span>
<span><span class="co">#&gt;  [91] evaluate_0.23          png_0.1-8              vctrs_0.6.5           </span></span>
<span><span class="co">#&gt;  [94] foreach_1.5.2          tidyr_1.3.1            gtable_0.3.4          </span></span>
<span><span class="co">#&gt;  [97] purrr_1.0.2            clue_0.3-65            ggplot2_3.5.0         </span></span>
<span><span class="co">#&gt; [100] cachem_1.0.8           xfun_0.42              broom_1.0.5           </span></span>
<span><span class="co">#&gt; [103] rstatix_0.7.2          ragg_1.3.0             tibble_3.2.1          </span></span>
<span><span class="co">#&gt; [106] iterators_1.0.14       AnnotationDbi_1.60.2   memoise_2.0.1         </span></span>
<span><span class="co">#&gt; [109] IRanges_2.32.0         cluster_2.1.6</span></span></code></pre></div>
</details>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>dominoSignal is developed by Christopher Cherry, Jacob T Mitchell, Sushma Nagaraj, Kavita Krishnan, Dmitrijs Lvovs.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
