<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="domino2">
<title>Get Started with domino2 • domino2</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<script src="../extra.js"></script><meta property="og:title" content="Get Started with domino2">
<meta property="og:description" content="domino2">
<meta property="og:image" content="/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-#7DDED2"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">domino2</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/domino2.html">Get started</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/domino_object_vignette.html">Interacting with domino2 Objects</a>
    <a class="dropdown-item" href="../articles/plotting_vignette.html">Plotting Functions and Options</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-news">News</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-news">
    <h6 class="dropdown-header" data-toc-skip>Releases</h6>
    <h6 class="dropdown-header" data-toc-skip>Version 0.2</h6>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../news/index.html">Changelog</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/FertigLab/domino_development/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="domino2_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Get Started with domino2</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/FertigLab/domino_development/blob/HEAD/vignettes/domino2.Rmd" class="external-link"><code>vignettes/domino2.Rmd</code></a></small>
      <div class="d-none name"><code>domino2.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="using-domino2-v0-2-2-for-cell-cell-communication-inference">Using domino2 v0.2.2 for cell-cell communication inference<a class="anchor" aria-label="anchor" href="#using-domino2-v0-2-2-for-cell-cell-communication-inference"></a>
</h2>
<p>domino2 is a tool for analysis of intra- and intercellular signaling in single cell RNA sequencing (scRNAseq) data based on transcription factor (TF) activation. Here, we show a basic example pipeline for generating communication networks as a domino object. This vignette will demonstrate how to use domino2 on the 10X Genomics Peripheral Blood Mononuclear Cells (PBMC) data set of 2,700 cells. The data can be downloaded <a href="https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz" class="external-link">here</a>.</p>
<div class="section level3">
<h3 id="preprocessing-scrnaseq-data">Preprocessing scRNAseq data<a class="anchor" aria-label="anchor" href="#preprocessing-scrnaseq-data"></a>
</h3>
<p>Analysis of cell-cell communication with domino2 often follows initial processing and annotation of scRNAseq data. We used the <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial" class="external-link">Satija Lab’s Guided Clustering Tutorial</a> to prepare the data set for analysis with domino2. The <a href="data-raw/example_data.R">complete processing script</a> is available in the data-raw directory of the domino2 package. The processed data can be downloaded from <a href="https://zenodo.org/records/10222767" class="external-link">Zenodo</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">pbmc_dir</span>, <span class="st">"/pbmc_seurat.rds"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="saving-preprocessed-data">Saving Preprocessed Data:<a class="anchor" aria-label="anchor" href="#saving-preprocessed-data"></a>
</h4>
<p>domino2 was designed to be compatible with <a href="https://satijalab.org/seurat" class="external-link">Seurat</a> objects in addition to accepting matrices and vectors of the data set as function inputs. Therefore, it’s a good idea to save the processed <a href="https://satijalab.org/seurat" class="external-link">Seurat</a> object for further analysis. However, <a href="https://pyscenic.readthedocs.io/en/latest/" class="external-link">pySCENIC</a> is used for TF activation inference, and it requires as input a cell by gene matrix. This is the opposite orientation of <a href="https://satijalab.org/seurat" class="external-link">Seurat</a> objects but is the default for Python based tools such as <a href="https://scanpy.readthedocs.io/en/stable/" class="external-link">scanpy</a>. The RNA counts matrix can be saved as a tab-seperated value (.tsv) or comma-sperated value (.csv) file after transposing the matrix to cell by gene orientation.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># save Seurat object as RDS</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">pbmc</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">input_dir</span>, <span class="st">"/pbmc.rda"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc_counts</span> <span class="op">&lt;-</span> <span class="fu">GetAssayData</span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, slot <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">pbmc_counts</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">input_dir</span>, <span class="st">"/pbmc3k_counts.tsv"</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"\t"</span>,</span>
<span>    col.names <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span></code></pre></div>
<p>tsv and csv files are very inefficient for storing large matrices. As an alternative, the counts matrix can be saved as a loom file and directly passed to <a href="https://pyscenic.readthedocs.io/en/latest/" class="external-link">pySCENIC</a> in this format. Generating a loom file in R requires use of the <a href="https://github.com/mojaveazure/loomR" class="external-link">loomR</a> package. The package automatically handles the transposition of the counts matrix to cell by gene orientation as well. We recommend this approach to passing a counts matrix to pySCENIC. However, users should be aware that <a href="https://github.com/mojaveazure/loomR" class="external-link">loomR</a> is not hosted on CRAN or BioConductor at the time of this vignette’s creation.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mojaveazure/loomR" class="external-link">loomR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># save loom counts matrix</span></span>
<span><span class="va">pbmc_counts</span> <span class="op">&lt;-</span> <span class="fu">GetAssayData</span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, slot <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></span>
<span><span class="va">pbmc_loom</span> <span class="op">&lt;-</span> <span class="fu">loomR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/loomR/man/create.html" class="external-link">create</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">input_dir</span>, <span class="st">"/pbmc3k_counts.loom"</span><span class="op">)</span>, data <span class="op">=</span> <span class="va">pbmc_counts</span><span class="op">)</span></span>
<span><span class="co"># connection to the loom file must be closed to complete file creation</span></span>
<span><span class="va">pbmc_loom</span><span class="op">$</span><span class="fu">close_all</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>The cell x gene counts matrix may also be saved as a tab-separated value (.tsv) file if <a href="https://github.com/mojaveazure/loomR" class="external-link">loomR</a> cannot be used on your computing resource, but be aware that dense formatting of a .tsv can lead to very large file sizes and slow processing for large data sets. We have provided the necessary code below, but due to the large file size, this .tsv file is not provided in our data directory.</p>
<p>If you would like to skip ahead to the rest of the analysis, the processed objects from these steps are available in the data directory of this package.</p>
</div>
</div>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>Installation of domino2 from Github can be achieved using the <a href="https://remotes.r-lib.org" class="external-link">remotes</a> package. The current stable version is v0.2.2, so we will install that branch.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://remotes.r-lib.org" class="external-link">remotes</a></span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"FertigLab/domino2@v0.2.2"</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="additional-downloads">Additional Downloads<a class="anchor" aria-label="anchor" href="#additional-downloads"></a>
</h4>
<p>For this tutorial, <a href="https://pyscenic.readthedocs.io/en/latest/" class="external-link">pySCENIC</a> is used as the method for TF activity inference and the assessment of ligand-receptor interactions is based on curated interactions from <a href="https://github.com/ventolab/cellphonedb-data/releases/tag/v4.0.0" class="external-link">CellPhoneDB v4.0.0</a>. Each of these requires downloading some files prior to use in our analysis pipeline.</p>
<div class="section level5">
<h5 id="downloads-for-scenic">Downloads for SCENIC:<a class="anchor" aria-label="anchor" href="#downloads-for-scenic"></a>
</h5>
<p>A singularity image of <a href="https://scenic.aertslab.org/" class="external-link">SCENIC</a> v0.12.1 can be installed from the DockerHub image as a source (example scripts for use with Docker are provided in the scenic_bash directory as well). <a href="https://scenic.aertslab.org/" class="external-link">SCENIC</a> requires a list of TFs, motif annotations, and cisTarget motifs which are all available from the authors of <a href="https://scenic.aertslab.org/" class="external-link">SCENIC</a> for human (HGNC), mouse (MGI), and fly. The following will download everything necessary for an analysis of a data set with HGNC gene labels for the hg38 genome.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1"></a><span class="va">SCENIC_DIR=</span><span class="st">"'temp_dir'/scenic"</span></span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="fu">mkdir</span> <span class="va">${SCENIC_DIR}</span></span>
<span id="cb5-3"><a href="#cb5-3"></a></span>
<span id="cb5-4"><a href="#cb5-4"></a><span class="co"># Build singularity image</span></span>
<span id="cb5-5"><a href="#cb5-5"></a><span class="ex">singularity</span> build <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/aertslab-pyscenic-0.12.1.sif"</span> docker://aertslab/pyscenic:0.12.1</span>
<span id="cb5-6"><a href="#cb5-6"></a></span>
<span id="cb5-7"><a href="#cb5-7"></a><span class="co"># Matrix containing motifs as rows and genes as columns and ranking position for each gene and motif (based on CRM scores) as values. URLs provided link to v2 feather files required for the 0.12.1 version of pySENIC.</span></span>
<span id="cb5-8"><a href="#cb5-8"></a></span>
<span id="cb5-9"><a href="#cb5-9"></a><span class="ex">curl</span> <span class="st">"https://resources.aertslab.org/cistarget/databases/homo_sapiens/hg38/refseq_r80/mc_v10_clust/gene_based/hg38_10kbp_up_10kbp_down_full_tx_v10_clust.genes_vs_motifs.rankings.feather"</span> \</span>
<span id="cb5-10"><a href="#cb5-10"></a>  -o <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/hg38_10kbp_up_10kbp_down_full_tx_v10_clust.genes_vs_motifs.rankings.feather"</span></span>
<span id="cb5-11"><a href="#cb5-11"></a></span>
<span id="cb5-12"><a href="#cb5-12"></a><span class="ex">curl</span> <span class="st">"https://resources.aertslab.org/cistarget/databases/homo_sapiens/hg38/refseq_r80/mc_v10_clust/gene_based/hg38_500bp_up_100bp_down_full_tx_v10_clust.genes_vs_motifs.rankings.feather"</span> \</span>
<span id="cb5-13"><a href="#cb5-13"></a>  -o <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/hg38_500bp_up_100bp_down_full_tx_v10_clust.genes_vs_motifs.rankings.feather"</span></span>
<span id="cb5-14"><a href="#cb5-14"></a></span>
<span id="cb5-15"><a href="#cb5-15"></a><span class="co"># List of genes encoding TFs in the HG38 reference genome</span></span>
<span id="cb5-16"><a href="#cb5-16"></a><span class="ex">curl</span> <span class="st">"https://resources.aertslab.org/cistarget/tf_lists/allTFs_hg38.txt"</span> \</span>
<span id="cb5-17"><a href="#cb5-17"></a>  -o <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/allTFs_hg38.txt"</span></span>
<span id="cb5-18"><a href="#cb5-18"></a></span>
<span id="cb5-19"><a href="#cb5-19"></a><span class="co"># Motif annotations based on the 2017 cisTarget motif collection. Use these files if you are using the mc9nr databases.</span></span>
<span id="cb5-20"><a href="#cb5-20"></a><span class="ex">curl</span> <span class="st">"https://resources.aertslab.org/cistarget/motif2tf/motifs-v9-nr.hgnc-m0.001-o0.0.tbl"</span> \</span>
<span id="cb5-21"><a href="#cb5-21"></a>  -o <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/motifs-v9-nr.hgnc-m0.001-o0.0.tbl"</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="downloads-for-cellphonedb">Downloads for CellPhoneDB:<a class="anchor" aria-label="anchor" href="#downloads-for-cellphonedb"></a>
</h5>
<p>Database files from <a href="https://github.com/ventolab/cellphonedb-data/releases/tag/v4.0.0" class="external-link">CellPhoneDB v4.0.0</a> for human scRNAseq data can be installed from a public Github repository from the Tiechmann Group that developed CellPhoneDB.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#</span></span>
<span><span class="va">cellphone_url</span> <span class="op">&lt;-</span> <span class="st">"https://github.com/ventolab/cellphonedb-data/archive/refs/tags/v4.0.0.tar.gz"</span></span>
<span></span>
<span><span class="co"># download compressed database</span></span>
<span><span class="va">cellphone_tar</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">temp_dir</span>, <span class="st">"/cellphoneDB_v4.tar.gz"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="va">cellphone_url</span>, destfile <span class="op">=</span> <span class="va">cellphone_tar</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># move contents of the compressed file to a new directory</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/untar.html" class="external-link">untar</a></span><span class="op">(</span>tarfile <span class="op">=</span> <span class="va">cellphone_tar</span>, exdir <span class="op">=</span> <span class="va">cellphone_dir</span><span class="op">)</span></span>
<span><span class="va">cellphone_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">cellphone_dir</span>, <span class="st">"/cellphonedb-data-4.0.0/data"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="scenic-analysis">SCENIC Analysis<a class="anchor" aria-label="anchor" href="#scenic-analysis"></a>
</h3>
<p>SCENIC is our lab’s preferred software for TF activity scoring. We recommend using the Python implementation (<a href="https://pyscenic.readthedocs.io/en/latest/" class="external-link">pySCENIC</a>) as it is faster than the original R implementation. Be aware, the use of <a href="https://scenic.aertslab.org/" class="external-link">SCENIC</a> is often the slowest and most memory intensive step of this analysis pipeline. <a href="https://scenic.aertslab.org/" class="external-link">SCENIC</a> should be run on computing resources with access to multi-core processing and large amounts of memory.</p>
<p>As described in the section on saving preprocessed data, <a href="https://scenic.aertslab.org/" class="external-link">SCENIC</a> requires RNA counts to be provided in a cell by gene matrix format, so we will be using the pbmc3k_counts.loom file in this stage of analysis.</p>
<p><a href="https://pyscenic.readthedocs.io/en/latest/" class="external-link">pySCENIC</a> is initiated using bash scripting in the terminal. The analysis consists of 3 steps to score genes for TF motif enrichment, construct TF regulons consisting of genes targeted by the TFs, and arrive at AUC scores for enrichment of regulon gene transcription within each cell.</p>
<div class="section level4">
<h4 id="grn-construct-tf-modules">grn: construct TF-modules<a class="anchor" aria-label="anchor" href="#grn-construct-tf-modules"></a>
</h4>
<p>Co-expression modules are used to quantify gene-TF adjacencies.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1"></a><span class="ex">singularity</span> exec <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/aertslab-pyscenic-0.12.1.sif"</span> pyscenic grn \</span>
<span id="cb7-2"><a href="#cb7-2"></a>    <span class="st">"../data/pbmc3k_counts.loom"</span> \</span>
<span id="cb7-3"><a href="#cb7-3"></a>    <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/allTFs_hg38.txt"</span> \</span>
<span id="cb7-4"><a href="#cb7-4"></a>    -o <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/pbmc_adj.tsv"</span> \</span>
<span id="cb7-5"><a href="#cb7-5"></a>    --num_workers 6 \</span>
<span id="cb7-6"><a href="#cb7-6"></a>    --seed 123</span>
<span id="cb7-7"><a href="#cb7-7"></a></span>
<span id="cb7-8"><a href="#cb7-8"></a><span class="co"># Arguments:</span></span>
<span id="cb7-9"><a href="#cb7-9"></a>    <span class="co"># path to the loom file</span></span>
<span id="cb7-10"><a href="#cb7-10"></a>    <span class="co"># list of TFs</span></span>
<span id="cb7-11"><a href="#cb7-11"></a>    <span class="co"># output directory for the adjacency matrix</span></span>
<span id="cb7-12"><a href="#cb7-12"></a>    <span class="co"># number of CPUs to use if multi-core processing is available</span></span>
<span id="cb7-13"><a href="#cb7-13"></a>    <span class="co"># specify a random seed for reproducibility</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="ctx-construct-tf-regulons-with-pruning-based-on-tf-motif-enrichment">ctx: construct TF regulons with pruning based on TF motif enrichment<a class="anchor" aria-label="anchor" href="#ctx-construct-tf-regulons-with-pruning-based-on-tf-motif-enrichment"></a>
</h4>
<p>The rankings of genes based on enrichment of TF motifs to the transcription start site (TSS) are considered in the construction of regulons, where target genes in the TF-modules are removed if they lack motifs proximal to the TSS where the TF may bind.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1"></a><span class="ex">singularity</span> exec <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/aertslab-pyscenic-0.12.1.sif"</span> pyscenic ctx \</span>
<span id="cb8-2"><a href="#cb8-2"></a>    <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/pbmc_adj.tsv"</span> \</span>
<span id="cb8-3"><a href="#cb8-3"></a>    <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/hg38_10kbp_up_10kbp_down_full_tx_v10_clust.genes_vs_motifs.rankings.feather"</span> \</span>
<span id="cb8-4"><a href="#cb8-4"></a>    <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/hg38_500bp_up_100bp_down_full_tx_v10_clust.genes_vs_motifs.rankings.feather"</span> \</span>
<span id="cb8-5"><a href="#cb8-5"></a>    --annotations_fname <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/motifs-v9-nr.hgnc-m0.001-o0.0.tbl"</span> \</span>
<span id="cb8-6"><a href="#cb8-6"></a>    --expression_mtx_fname <span class="st">"data/pbmc3k_counts.loom"</span> \</span>
<span id="cb8-7"><a href="#cb8-7"></a>    --mode <span class="st">"dask_multiprocessing"</span> \</span>
<span id="cb8-8"><a href="#cb8-8"></a>    --output <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/pbmc_regulons.csv"</span> \</span>
<span id="cb8-9"><a href="#cb8-9"></a>    --num_workers 1                                                          </span>
<span id="cb8-10"><a href="#cb8-10"></a></span>
<span id="cb8-11"><a href="#cb8-11"></a><span class="co"># Arguments:</span></span>
<span id="cb8-12"><a href="#cb8-12"></a>    <span class="co"># adjacency matrix output from grn</span></span>
<span id="cb8-13"><a href="#cb8-13"></a>    <span class="co"># target rankings of motif enrichment within 10 kb of TSS</span></span>
<span id="cb8-14"><a href="#cb8-14"></a>    <span class="co"># target rankings of motif enrichment within 500 bp upstream and 100 bp downstream of TSS</span></span>
<span id="cb8-15"><a href="#cb8-15"></a>    <span class="co"># TF motif features</span></span>
<span id="cb8-16"><a href="#cb8-16"></a>    <span class="co"># counts matrix loom file</span></span>
<span id="cb8-17"><a href="#cb8-17"></a>    <span class="co"># enable multi-core processing</span></span>
<span id="cb8-18"><a href="#cb8-18"></a>    <span class="co"># output file of learned TF regulons</span></span>
<span id="cb8-19"><a href="#cb8-19"></a>    <span class="co"># number of CPU cores</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="aucell-calculate-tf-activity-scores">aucell: calculate TF activity scores<a class="anchor" aria-label="anchor" href="#aucell-calculate-tf-activity-scores"></a>
</h4>
<p>Enrichment of a regulon is measured as the Area Under the recovery Curve (AUC) of the genes that define this regulon.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1"></a><span class="ex">singularity</span> exec <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/aertslab-pyscenic-0.12.1.sif"</span> pyscenic aucell \</span>
<span id="cb9-2"><a href="#cb9-2"></a>    <span class="st">"../data/pbmc3k_counts.loom"</span> \</span>
<span id="cb9-3"><a href="#cb9-3"></a>    <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/pbmc_regulons.csv"</span> \</span>
<span id="cb9-4"><a href="#cb9-4"></a>    -o <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/pbmc_auc.csv"</span></span>
<span id="cb9-5"><a href="#cb9-5"></a>    </span>
<span id="cb9-6"><a href="#cb9-6"></a><span class="co"># Arguments:</span></span>
<span id="cb9-7"><a href="#cb9-7"></a>    <span class="co"># counts matrix loom file</span></span>
<span id="cb9-8"><a href="#cb9-8"></a>    <span class="co"># regulon table output from ctx</span></span>
<span id="cb9-9"><a href="#cb9-9"></a>    <span class="co"># cell x TF matrix of TF enrichment AUC values</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="building-a-signaling-network-with-domino2">Building a Signaling Network with domino2<a class="anchor" aria-label="anchor" href="#building-a-signaling-network-with-domino2"></a>
</h3>
<div class="section level4">
<h4 id="loading-scenic-results">Loading SCENIC Results<a class="anchor" aria-label="anchor" href="#loading-scenic-results"></a>
</h4>
<p>The TF activities are a required input for domino2, and the regulons learned by <a href="https://scenic.aertslab.org/" class="external-link">SCENIC</a> are an important but optional input needed to annotate TF-target interactions and to prune TF-receptor linkages where the receptor is a target of the TF. This prevents the distinction of receptor expression driving TF activity or the TF inducing the receptor’s expression.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">regulons</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">scenic_dir</span>, <span class="st">"/regulons_pbmc_3k.csv"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">auc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">scenic_dir</span>, <span class="st">"/auc_pbmc_3k.csv"</span><span class="op">)</span>, header <span class="op">=</span> <span class="cn">TRUE</span>, row.names <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span>, sep <span class="op">=</span> <span class="st">","</span><span class="op">)</span></span></code></pre></div>
<p>The initial regulons data frame read into R from the ctx function has 2 rows of column names that need to be replaced with one succinct description. domino2 has changed the input format for TF regulons to be a list storing vectors of target genes in each regulon, where the names of the list are the TF genes. This facilitates the use of alternative methods for TF activity quantification. We provide a helper function, <code><a href="../reference/create_regulon_list_scenic.html">create_regulon_list_scenic()</a></code> for easy retrieval of TF regulons from the output of the <a href="https://pyscenic.readthedocs.io/en/latest/" class="external-link">pySCENIC</a> ctx function.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">regulons</span> <span class="op">&lt;-</span> <span class="va">regulons</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">:</span><span class="op">-</span><span class="fl">2</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">regulons</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TF"</span>, <span class="st">"MotifID"</span>, <span class="st">"AUC"</span>, <span class="st">"NES"</span>, <span class="st">"MotifSimilarityQvalue"</span>, <span class="st">"OrthologousIdentity"</span>,</span>
<span>    <span class="st">"Annotation"</span>, <span class="st">"Context"</span>, <span class="st">"TargetGenes"</span>, <span class="st">"RankAtMax"</span><span class="op">)</span></span>
<span><span class="va">regulon_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_regulon_list_scenic.html">create_regulon_list_scenic</a></span><span class="op">(</span>regulons <span class="op">=</span> <span class="va">regulons</span><span class="op">)</span></span></code></pre></div>
<p>Users should be aware that the AUC matrix from <a href="https://scenic.aertslab.org/" class="external-link">SCENIC</a> is loaded in a cell x TF orientation and should be transposed to TF x cell orientation. pySCENIC also appends “(+)” to TF names that are converted to “…” upon loading into R. These characters can be included without affecting the results of domino2 analysis, but can be confusing when querying TF features in the data. We recommend comprehensive removal of the “…” characters using the <code><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub()</a></code> function.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">auc_in</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">auc</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Remove pattern '...' from the end of all rownames:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">auc_in</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"\\.\\.\\.$"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">auc_in</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="load-cellphonedb-database">Load CellPhoneDB Database<a class="anchor" aria-label="anchor" href="#load-cellphonedb-database"></a>
</h4>
<p>domino2 has been updated to read ligand-receptor data bases in a uniform data.frame format referred to as a receptor-ligand map (rl_map) to enable the use of alternative or updated reference databases as opposed to a particular version of CellPhoneDB’s database in older versions of Domino. Each row corresponds to a ligand-receptor interaction. Genes participating in an interaction are referred to as “partner A” and “partner B” without a requirement for fixed ordering of whether A or B is the ligand and vice versa. The minimum required columns of this data frame are:</p>
<ul>
<li>int_pair: the names of the interacting ligand and receptor separated by " &amp; "</li>
<li>gene_A: the gene or genes encoding partner A</li>
<li>gene_B: the gene or genes encoding partner B</li>
<li>type_A: (“L”, “R”) - indicates whether partner A is a ligand (“L”) or receptor (“R”)</li>
<li>type_B: (“L”, “R”) - indicates whether partner B is a ligand (“L”) or receptor (“R”)</li>
</ul>
<p>Additional annotation columns can be provided such as name_A and name_B for ligands or receptors whose name in the interaction database does not match the names of their encoding genes. This formatting also allows for the consideration of ligand and receptor complexes comprised of a heteromeric combination of multiple proteins that must be co-expressed to function. In these cases, the “name_*” column shows the name of the protein complex, and the “gene_*” column shows the names of the genes encoding the components of the complex separated by commas “,”. When plotting results from the build domino object, the names of the interacting ligands and receptors will be used based on combinatorial expression of the complex components.</p>
<p>To facilitate the use of this formatting with the CellPhoneDB database, we include a helper function, <code><a href="../reference/create_rl_map_cellphonedb.html">create_rl_map_cellphonedb()</a></code>, that automatically parses files from the cellPhoneDB database to arrive at the rl_map format.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">complexes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">cellphone_dir</span>, <span class="st">"/cellphonedb-data-4.0.0/data/complex_input.csv"</span><span class="op">)</span>,</span>
<span>    stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">cellphone_dir</span>, <span class="st">"/cellphonedb-data-4.0.0/data/gene_input.csv"</span><span class="op">)</span>,</span>
<span>    stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">interactions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">cellphone_dir</span>, <span class="st">"/cellphonedb-data-4.0.0/data/interaction_input.csv"</span><span class="op">)</span>,</span>
<span>    stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">proteins</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">cellphone_dir</span>, <span class="st">"/cellphonedb-data-4.0.0/data/protein_input.csv"</span><span class="op">)</span>,</span>
<span>    stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">rl_map</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_rl_map_cellphonedb.html">create_rl_map_cellphonedb</a></span><span class="op">(</span>genes <span class="op">=</span> <span class="va">genes</span>, proteins <span class="op">=</span> <span class="va">proteins</span>, interactions <span class="op">=</span> <span class="va">interactions</span>,</span>
<span>    complexes <span class="op">=</span> <span class="va">complexes</span>, database_name <span class="op">=</span> <span class="st">"CellPhoneDB_v4.0"</span>  <span class="co"># database version used</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">rl_map</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="1%">
<col width="21%">
<col width="8%">
<col width="7%">
<col width="6%">
<col width="3%">
<col width="12%">
<col width="5%">
<col width="4%">
<col width="3%">
<col width="10%">
<col width="8%">
<col width="8%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="left">int_pair</th>
<th align="left">name_A</th>
<th align="left">uniprot_A</th>
<th align="left">gene_A</th>
<th align="left">type_A</th>
<th align="left">name_B</th>
<th align="left">uniprot_B</th>
<th align="left">gene_B</th>
<th align="left">type_B</th>
<th align="left">annotation_strategy</th>
<th align="left">source</th>
<th align="left">database_name</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">4</td>
<td align="left">RAreceptor_RXRG &amp; atRetinoicAcid_byALDH1A3</td>
<td align="left">RAreceptor_RXRG</td>
<td align="left">P48443,P29373</td>
<td align="left">RXRG,CRABP2</td>
<td align="left">R</td>
<td align="left">atRetinoicAcid_byALDH1A3</td>
<td align="left">P47895</td>
<td align="left">ALDH1A3</td>
<td align="left">L</td>
<td align="left">curated</td>
<td align="left">uniprot;reactome</td>
<td align="left">CellPhoneDB_v4.0</td>
</tr>
<tr class="even">
<td align="left">5</td>
<td align="left">RAreceptor_RXRG &amp; atRetinoicAcid_byALDH1A2</td>
<td align="left">RAreceptor_RXRG</td>
<td align="left">P48443,P29373</td>
<td align="left">RXRG,CRABP2</td>
<td align="left">R</td>
<td align="left">atRetinoicAcid_byALDH1A2</td>
<td align="left">O94788</td>
<td align="left">ALDH1A2</td>
<td align="left">L</td>
<td align="left">curated</td>
<td align="left">uniprot;reactome</td>
<td align="left">CellPhoneDB_v4.0</td>
</tr>
<tr class="odd">
<td align="left">6</td>
<td align="left">RAreceptor_RXRG &amp; atRetinoicAcid_byALDH1A1</td>
<td align="left">RAreceptor_RXRG</td>
<td align="left">P48443,P29373</td>
<td align="left">RXRG,CRABP2</td>
<td align="left">R</td>
<td align="left">atRetinoicAcid_byALDH1A1</td>
<td align="left">P00352</td>
<td align="left">ALDH1A1</td>
<td align="left">L</td>
<td align="left">curated</td>
<td align="left">uniprot;reactome</td>
<td align="left">CellPhoneDB_v4.0</td>
</tr>
<tr class="even">
<td align="left">7</td>
<td align="left">RAreceptor_RXRB &amp; atRetinoicAcid_byALDH1A3</td>
<td align="left">RAreceptor_RXRB</td>
<td align="left">P28702,P29373</td>
<td align="left">RXRB,CRABP2</td>
<td align="left">R</td>
<td align="left">atRetinoicAcid_byALDH1A3</td>
<td align="left">P47895</td>
<td align="left">ALDH1A3</td>
<td align="left">L</td>
<td align="left">curated</td>
<td align="left">uniprot;reactome</td>
<td align="left">CellPhoneDB_v4.0</td>
</tr>
<tr class="odd">
<td align="left">8</td>
<td align="left">RAreceptor_RXRB &amp; atRetinoicAcid_byALDH1A2</td>
<td align="left">RAreceptor_RXRB</td>
<td align="left">P28702,P29373</td>
<td align="left">RXRB,CRABP2</td>
<td align="left">R</td>
<td align="left">atRetinoicAcid_byALDH1A2</td>
<td align="left">O94788</td>
<td align="left">ALDH1A2</td>
<td align="left">L</td>
<td align="left">curated</td>
<td align="left">uniprot;reactome</td>
<td align="left">CellPhoneDB_v4.0</td>
</tr>
<tr class="even">
<td align="left">9</td>
<td align="left">RAreceptor_RXRB &amp; atRetinoicAcid_byALDH1A1</td>
<td align="left">RAreceptor_RXRB</td>
<td align="left">P28702,P29373</td>
<td align="left">RXRB,CRABP2</td>
<td align="left">R</td>
<td align="left">atRetinoicAcid_byALDH1A1</td>
<td align="left">P00352</td>
<td align="left">ALDH1A1</td>
<td align="left">L</td>
<td align="left">curated</td>
<td align="left">uniprot;reactome</td>
<td align="left">CellPhoneDB_v4.0</td>
</tr>
</tbody>
</table>
<div class="section level5">
<h5 id="optional-adding-interactions-manually">Optional: Adding interactions manually<a class="anchor" aria-label="anchor" href="#optional-adding-interactions-manually"></a>
</h5>
<p>The change to use rl_map formatting also enables users to manually append interactions of interest that are not included in the interaction database if need be. This can be attained by formatting the desired interactions as a data.frame with the same column headers as the rl_map and using the <code><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind()</a></code> function.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Integrin complexes are not annotated as receptors in CellPhoneDB_v4.0</span></span>
<span><span class="co"># collagen-integrin interactions between cells may be missed unless tables from</span></span>
<span><span class="co"># the CellPhoneDB reference are edited or the interactions are manually added</span></span>
<span></span>
<span><span class="va">col_int_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>int_pair <span class="op">=</span> <span class="st">"a11b1 complex &amp; COLA1_HUMAN"</span>, name_A <span class="op">=</span> <span class="st">"a11b1 complex"</span>,</span>
<span>    uniprot_A <span class="op">=</span> <span class="st">"P05556,Q9UKX5"</span>, gene_A <span class="op">=</span> <span class="st">"ITB1,ITA11"</span>, type_A <span class="op">=</span> <span class="st">"R"</span>, name_B <span class="op">=</span> <span class="st">"COLA1_HUMAN"</span>,</span>
<span>    uniprot_B <span class="op">=</span> <span class="st">"P02452,P08123"</span>, gene_B <span class="op">=</span> <span class="st">"COL1A1,COL1A2"</span>, type_B <span class="op">=</span> <span class="st">"L"</span>, annotation_strategy <span class="op">=</span> <span class="st">"manual"</span>,</span>
<span>    source <span class="op">=</span> <span class="st">"manual"</span>, database_name <span class="op">=</span> <span class="st">"manual"</span><span class="op">)</span></span>
<span><span class="va">rl_map_append</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">col_int_df</span>, <span class="va">rl_map</span><span class="op">)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">rl_map_append</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="1%">
<col width="20%">
<col width="7%">
<col width="6%">
<col width="5%">
<col width="3%">
<col width="11%">
<col width="6%">
<col width="6%">
<col width="3%">
<col width="9%">
<col width="8%">
<col width="8%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="left">int_pair</th>
<th align="left">name_A</th>
<th align="left">uniprot_A</th>
<th align="left">gene_A</th>
<th align="left">type_A</th>
<th align="left">name_B</th>
<th align="left">uniprot_B</th>
<th align="left">gene_B</th>
<th align="left">type_B</th>
<th align="left">annotation_strategy</th>
<th align="left">source</th>
<th align="left">database_name</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">1</td>
<td align="left">a11b1 complex &amp; COLA1_HUMAN</td>
<td align="left">a11b1 complex</td>
<td align="left">P05556,Q9UKX5</td>
<td align="left">ITB1,ITA11</td>
<td align="left">R</td>
<td align="left">COLA1_HUMAN</td>
<td align="left">P02452,P08123</td>
<td align="left">COL1A1,COL1A2</td>
<td align="left">L</td>
<td align="left">manual</td>
<td align="left">manual</td>
<td align="left">manual</td>
</tr>
<tr class="even">
<td align="left">4</td>
<td align="left">RAreceptor_RXRG &amp; atRetinoicAcid_byALDH1A3</td>
<td align="left">RAreceptor_RXRG</td>
<td align="left">P48443,P29373</td>
<td align="left">RXRG,CRABP2</td>
<td align="left">R</td>
<td align="left">atRetinoicAcid_byALDH1A3</td>
<td align="left">P47895</td>
<td align="left">ALDH1A3</td>
<td align="left">L</td>
<td align="left">curated</td>
<td align="left">uniprot;reactome</td>
<td align="left">CellPhoneDB_v4.0</td>
</tr>
<tr class="odd">
<td align="left">5</td>
<td align="left">RAreceptor_RXRG &amp; atRetinoicAcid_byALDH1A2</td>
<td align="left">RAreceptor_RXRG</td>
<td align="left">P48443,P29373</td>
<td align="left">RXRG,CRABP2</td>
<td align="left">R</td>
<td align="left">atRetinoicAcid_byALDH1A2</td>
<td align="left">O94788</td>
<td align="left">ALDH1A2</td>
<td align="left">L</td>
<td align="left">curated</td>
<td align="left">uniprot;reactome</td>
<td align="left">CellPhoneDB_v4.0</td>
</tr>
<tr class="even">
<td align="left">6</td>
<td align="left">RAreceptor_RXRG &amp; atRetinoicAcid_byALDH1A1</td>
<td align="left">RAreceptor_RXRG</td>
<td align="left">P48443,P29373</td>
<td align="left">RXRG,CRABP2</td>
<td align="left">R</td>
<td align="left">atRetinoicAcid_byALDH1A1</td>
<td align="left">P00352</td>
<td align="left">ALDH1A1</td>
<td align="left">L</td>
<td align="left">curated</td>
<td align="left">uniprot;reactome</td>
<td align="left">CellPhoneDB_v4.0</td>
</tr>
<tr class="odd">
<td align="left">7</td>
<td align="left">RAreceptor_RXRB &amp; atRetinoicAcid_byALDH1A3</td>
<td align="left">RAreceptor_RXRB</td>
<td align="left">P28702,P29373</td>
<td align="left">RXRB,CRABP2</td>
<td align="left">R</td>
<td align="left">atRetinoicAcid_byALDH1A3</td>
<td align="left">P47895</td>
<td align="left">ALDH1A3</td>
<td align="left">L</td>
<td align="left">curated</td>
<td align="left">uniprot;reactome</td>
<td align="left">CellPhoneDB_v4.0</td>
</tr>
<tr class="even">
<td align="left">8</td>
<td align="left">RAreceptor_RXRB &amp; atRetinoicAcid_byALDH1A2</td>
<td align="left">RAreceptor_RXRB</td>
<td align="left">P28702,P29373</td>
<td align="left">RXRB,CRABP2</td>
<td align="left">R</td>
<td align="left">atRetinoicAcid_byALDH1A2</td>
<td align="left">O94788</td>
<td align="left">ALDH1A2</td>
<td align="left">L</td>
<td align="left">curated</td>
<td align="left">uniprot;reactome</td>
<td align="left">CellPhoneDB_v4.0</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div class="section level3">
<h3 id="creating-the-domino-object">Creating the Domino Object<a class="anchor" aria-label="anchor" href="#creating-the-domino-object"></a>
</h3>
<p>domino2 analysis takes place in two steps. 1. <code><a href="../reference/create_domino.html">create_domino()</a></code></p>
<ol style="list-style-type: lower-alpha">
<li><p>initializes the domino result object</p></li>
<li><p>assesses differential TF activity across cell clusters by wilcoxon rank-sum test</p></li>
<li><p>establishes TF-receptor linkages based on Spearman correlation of TF activities with receptor expression across the queried data set.</p></li>
</ol>
<ol start="2" style="list-style-type: decimal">
<li><code><a href="../reference/build_domino.html">build_domino()</a></code></li>
</ol>
<ol style="list-style-type: lower-alpha">
<li><p>sets the parameters for which TFs and receptors are called as active within a cell cluster</p></li>
<li><p>aggregates the scaled expression of ligands capable of interacting with the active receptors for assessment of ligand type and cellular source triggering activation of a receptor.</p></li>
</ol>
<div class="section level4">
<h4 id="required-inputs-from-data-set">Required inputs from data set<a class="anchor" aria-label="anchor" href="#required-inputs-from-data-set"></a>
</h4>
<p>domino2 infers active receipt of signals via receptors based on the correlation of the receptor’s expression with TF activity across the data set and differential activity of the TF within a cell cluster. Correlations are conducted using scaled expression values rather than raw counts or normalized counts. For assessment of receptor activity on a per cell type basis, a named vector of cell cluster assignments, where the names are cell barcodes matching the expression matrix, are provided. Assessing signaling based on other categorical groupings of cells can be achieved by passing these groupings as “clusters” to <code><a href="../reference/build_domino.html">build_domino()</a></code> in place of cell types. domino2 accepts either a <a href="https://satijalab.org/seurat" class="external-link">Seurat</a> object with counts, scaled counts, and clusters included, or it requires a matrix of counts, a matrix of scaled counts, and a named vector of cell cluster labels as factors. Shown below is how to extract these elements from a <a href="https://satijalab.org/seurat" class="external-link">Seurat</a> object.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu">GetAssayData</span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, slot <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></span>
<span><span class="va">z_scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu">GetAssayData</span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, slot <span class="op">=</span> <span class="st">"scale.data"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cell_type</span></span></code></pre></div>
<p>Note: Ligand and receptor expression can only be assessed for genes included in the z_scores matrix. Many scRNAseq analysis pipelines recommend storing only genes with high variance in scaled expression slots for these data objects, thereby missing many genes encoding ligands and receptors. <strong>Ensure that all your genes of interest are included in the rows of your z_scores matrix.</strong> Scaled expression was calculated for all genes in this PBMC data set after removal of genes expressed in less than 3 cells.</p>
</div>
<div class="section level4">
<h4 id="create-domino-object">Create Domino object<a class="anchor" aria-label="anchor" href="#create-domino-object"></a>
</h4>
<p>At this point, the <code><a href="../reference/create_domino.html">create_domino()</a></code> function can be used to make the object. Parameters of note include “use_clusters” which is required to assess signaling between cell types rather than linkage of TFs and receptors broadly across the data set. “use_complexes” decides if receptors that function in heteromeric complexes will be considered in testing linkages between TFs and receptors. If TRUE, a receptor complex is only linked to a TF if a majority of the component genes meet the Spearman correlation threshold. “remove_rec_dropout” decides whether receptor values of 0 will be considered in correlation calculations; this measure is intended to reduce the effect of dropout on receptors with low expression.</p>
<p>To run <code><a href="../reference/create_domino.html">create_domino()</a></code> with matrix and vector inputs:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_dom</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_domino.html">create_domino</a></span><span class="op">(</span>rl_map <span class="op">=</span> <span class="va">rl_map</span>, features <span class="op">=</span> <span class="va">auc_in</span>, counts <span class="op">=</span> <span class="va">counts</span>, z_scores <span class="op">=</span> <span class="va">z_scores</span>,</span>
<span>    clusters <span class="op">=</span> <span class="va">clusters</span>, tf_targets <span class="op">=</span> <span class="va">regulon_list</span>, use_clusters <span class="op">=</span> <span class="cn">TRUE</span>, use_complexes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    remove_rec_dropout <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co"># rl_map: receptor-ligand map data frame features: TF scores (AUC matrix)</span></span>
<span><span class="co"># counts: counts matrix z_scores: scaled expression data clusters: named vector</span></span>
<span><span class="co"># of cell cluster assignments tf_targets: list of TFs and their regulons</span></span>
<span><span class="co"># use_clusters: assess receptor activation and ligand expression on a</span></span>
<span><span class="co"># per-cluster basis use_complexes: include receptors and genes that function as</span></span>
<span><span class="co"># a complex in results remove_rec_dropout: whether to remove zeroes from</span></span>
<span><span class="co"># correlation calculations</span></span></code></pre></div>
<p>The above results will be equivalent to using the <a href="https://satijalab.org/seurat" class="external-link">Seurat</a> object directly. Note that inclusion of both the matrix inputs and the <a href="https://satijalab.org/seurat" class="external-link">Seurat</a> object will prioritize the <a href="https://satijalab.org/seurat" class="external-link">Seurat</a> object.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_dom</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_domino.html">create_domino</a></span><span class="op">(</span>rl_map <span class="op">=</span> <span class="va">rl_map</span>, features <span class="op">=</span> <span class="va">auc_in</span>, ser <span class="op">=</span> <span class="va">pbmc</span>, tf_targets <span class="op">=</span> <span class="va">regulon_list</span>,</span>
<span>    use_clusters <span class="op">=</span> <span class="cn">TRUE</span>, use_complexes <span class="op">=</span> <span class="cn">TRUE</span>, remove_rec_dropout <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>For information on what is stored in a domino object and how to access it, please see our <a href="vignette(%22domino_object_vignette%22)">vignette on the structure of domino objects</a>.</p>
</div>
<div class="section level4">
<h4 id="build-domino">Build Domino<a class="anchor" aria-label="anchor" href="#build-domino"></a>
</h4>
<p><code><a href="../reference/build_domino.html">build_domino()</a></code> finalizes the construction of the domino object by setting parameters for identifying TFs with differential activation between clusters, receptor linkage with TFs based on magnitude of positive correlation, and the minimum percentage of cells within a cluster that have expression of a receptor for the receptor to be called as active.</p>
<p>There are also options for thresholds of the number of TFs that may be called active in a cluster and the number of receptors that may be linked to any one TF. For thresholds of <em>n</em> TFs and <em>m</em> receptors, the bottom <em>n</em> TFs by lowest p-values from the wilcoxon rank sum test and the top <em>m</em> receptors by Spearman correlation coefficient are chosen.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_dom</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/build_domino.html">build_domino</a></span><span class="op">(</span>dom <span class="op">=</span> <span class="va">pbmc_dom</span>, min_tf_pval <span class="op">=</span> <span class="fl">0.001</span>, max_tf_per_clust <span class="op">=</span> <span class="fl">25</span>,</span>
<span>    max_rec_per_tf <span class="op">=</span> <span class="fl">25</span>, rec_tf_cor_threshold <span class="op">=</span> <span class="fl">0.25</span>, min_rec_percentage <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span><span class="co"># min_tf_pval: Threshold for p-value of DE for TFs rec_tf_cor_threshold:</span></span>
<span><span class="co"># Minimum correlation between receptor and TF min_rec_percentage: Minimum</span></span>
<span><span class="co"># percent of cells that must express receptor</span></span></code></pre></div>
<p>Both of the thresholds for the number of receptors and TFs can be sent to infinity (Inf) to collect all receptors and TFs that meet statistical significance thresholds.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_dom_all</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/build_domino.html">build_domino</a></span><span class="op">(</span>dom <span class="op">=</span> <span class="va">pbmc_dom</span>, min_tf_pval <span class="op">=</span> <span class="fl">0.001</span>, max_tf_per_clust <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>    max_rec_per_tf <span class="op">=</span> <span class="cn">Inf</span>, rec_tf_cor_threshold <span class="op">=</span> <span class="fl">0.25</span>, min_rec_percentage <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="visualization-of-domino-results">Visualization of Domino Results<a class="anchor" aria-label="anchor" href="#visualization-of-domino-results"></a>
</h3>
<p>Multiple functions are available to visualize the intracellular networks between receptors and TFs and the ligand-receptor mediated intercellular networks between cell types.</p>
<div class="section level4">
<h4 id="summarize-tf-activity-and-linkage">Summarize TF Activity and Linkage<a class="anchor" aria-label="anchor" href="#summarize-tf-activity-and-linkage"></a>
</h4>
<p>Enrichment of TF activities by cell types can be visualized by <code><a href="../reference/feat_heatmap.html">feat_heatmap()</a></code> which plots data set-wide TF activity scores as a heatmap.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/feat_heatmap.html">feat_heatmap</a></span><span class="op">(</span><span class="va">pbmc_dom</span>, norm <span class="op">=</span> <span class="cn">TRUE</span>, bool <span class="op">=</span> <span class="cn">FALSE</span>, use_raster <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="domino2_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="cumulative-signaling-between-cell-types">Cumulative signaling between cell types<a class="anchor" aria-label="anchor" href="#cumulative-signaling-between-cell-types"></a>
</h4>
<p>The cumulative degree of signaling between clusters is assessed as the sum of the scaled expression of ligands targeting active receptors on another cluster. This can be visualized in graph format using the <code><a href="../reference/signaling_network.html">signaling_network()</a></code> function. Nodes represent each cell cluster and edges scale with the magnitude of signaling between the clusters. The color of the edge corresponds to the sender cluster for that signal.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/signaling_network.html">signaling_network</a></span><span class="op">(</span><span class="va">pbmc_dom</span>, edge_weight <span class="op">=</span> <span class="fl">1</span>, max_thresh <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p><img src="domino2_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>Signaling networks can also be drawn with the edges only rendering the signals directed towards a given cell type or signals from one cell type directed to others. To see those options in use, as well as other plotting functions and their options, please see our <a href="vignette(%22plotting_vignette%22)">plotting vignette</a>.</p>
</div>
<div class="section level4">
<h4 id="specific-signaling-interactions-between-clusters">Specific Signaling Interactions between Clusters<a class="anchor" aria-label="anchor" href="#specific-signaling-interactions-between-clusters"></a>
</h4>
<p>Beyond the aggregated degree of signaling between cell types, the degrees of signaling through specific ligand-receptor interactions can be assessed. <code><a href="../reference/gene_network.html">gene_network()</a></code> provides a graph of linkages between active TFs in a cluster, the linked receptors in that cluster, and the possible ligands of these active receptors.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gene_network.html">gene_network</a></span><span class="op">(</span><span class="va">pbmc_dom</span>, clust <span class="op">=</span> <span class="st">"dendritic_cell"</span>, layout <span class="op">=</span> <span class="st">"grid"</span><span class="op">)</span></span></code></pre></div>
<p><img src="domino2_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>New to domino2, <code><a href="../reference/gene_network.html">gene_network()</a></code> can be used between two clusters to determine if any of the possible ligands for a given receptor are expressed by a putative outgoing signaling cluster.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gene_network.html">gene_network</a></span><span class="op">(</span><span class="va">pbmc_dom</span>, clust <span class="op">=</span> <span class="st">"dendritic_cell"</span>, OutgoingSignalingClust <span class="op">=</span> <span class="st">"CD14_monocyte"</span>,</span>
<span>    layout <span class="op">=</span> <span class="st">"grid"</span><span class="op">)</span></span></code></pre></div>
<p><img src="domino2_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>A comprehensive assessment of ligand expression targeting active receptors on a given cluster can be assessed with <code><a href="../reference/incoming_signaling_heatmap.html">incoming_signaling_heatmap()</a></code>.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/incoming_signaling_heatmap.html">incoming_signaling_heatmap</a></span><span class="op">(</span><span class="va">pbmc_dom</span>, rec_clust <span class="op">=</span> <span class="st">"dendritic_cell"</span>, max_thresh <span class="op">=</span> <span class="fl">2.5</span>,</span>
<span>    use_raster <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="domino2_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>Another form of comprehensive ligand expression assessment is available for individual active receptors in the form of circos plots new to domino2. The outer arcs correspond to clusters in the domino object with inner arcs representing each possible ligand of the plotted receptor. Arcs are drawn between ligands on a cell type and the receptor if the ligand is expressed above the specified threshold. Arc widths correspond to the mean express of the ligand by the cluster with the widest arc width scaling to the maximum expression of the ligand within the data.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/circos_ligand_receptor.html">circos_ligand_receptor</a></span><span class="op">(</span><span class="va">pbmc_dom</span>, receptor <span class="op">=</span> <span class="st">"CD74"</span><span class="op">)</span></span></code></pre></div>
<p><img src="domino2_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
</div>
<div class="section level3">
<h3 id="continued-development">Continued Development<a class="anchor" aria-label="anchor" href="#continued-development"></a>
</h3>
<p>Since domino2 is a package still being developed, there are new functions and features that will be implemented in future versions. In the meantime, we have put together further information on <a href="vignette(%22plotting_vignette%22)">plotting</a> and the <a href="vignette(%22domino_object_vignette%22)">domino object structure</a> if you would like to explore more of the package’s functionality. Additionally, if you find any bugs, have further questions, or want to share an idea, please let us know <a href="https://github.com/FertigLab/domino_development/issues" class="external-link">here</a>.</p>
<details><p><summary>Vignette Build Information</summary> Date last built and session information:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.Date</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "2023-12-13"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0</span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] grid      stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] knitr_1.45            ComplexHeatmap_2.14.0 circlize_0.4.15      </span></span>
<span><span class="co">#&gt; [4] plyr_1.8.9            Seurat_5.0.1          SeuratObject_5.0.1   </span></span>
<span><span class="co">#&gt; [7] sp_2.1-1              domino2_0.2.2        </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] utf8_1.2.4             spatstat.explore_3.2-5 reticulate_1.34.0     </span></span>
<span><span class="co">#&gt;   [4] tidyselect_1.2.0       RSQLite_2.3.3          AnnotationDbi_1.60.2  </span></span>
<span><span class="co">#&gt;   [7] htmlwidgets_1.6.2      Rtsne_0.16             munsell_0.5.0         </span></span>
<span><span class="co">#&gt;  [10] codetools_0.2-19       ragg_1.2.6             ica_1.0-3             </span></span>
<span><span class="co">#&gt;  [13] future_1.33.0          miniUI_0.1.1.1         spatstat.random_3.2-1 </span></span>
<span><span class="co">#&gt;  [16] colorspace_2.1-0       progressr_0.14.0       Biobase_2.58.0        </span></span>
<span><span class="co">#&gt;  [19] filelock_1.0.2         highr_0.10             stats4_4.2.1          </span></span>
<span><span class="co">#&gt;  [22] ROCR_1.0-11            tensor_1.5             ggsignif_0.6.4        </span></span>
<span><span class="co">#&gt;  [25] listenv_0.9.0          GenomeInfoDbData_1.2.9 polyclip_1.10-6       </span></span>
<span><span class="co">#&gt;  [28] bit64_4.0.5            rprojroot_2.0.4        parallelly_1.36.0     </span></span>
<span><span class="co">#&gt;  [31] vctrs_0.6.4            generics_0.1.3         xfun_0.41             </span></span>
<span><span class="co">#&gt;  [34] BiocFileCache_2.6.1    R6_2.5.1               doParallel_1.0.17     </span></span>
<span><span class="co">#&gt;  [37] GenomeInfoDb_1.34.9    clue_0.3-65            bitops_1.0-7          </span></span>
<span><span class="co">#&gt;  [40] spatstat.utils_3.0-4   cachem_1.0.8           promises_1.2.1        </span></span>
<span><span class="co">#&gt;  [43] scales_1.3.0           gtable_0.3.4           Cairo_1.6-1           </span></span>
<span><span class="co">#&gt;  [46] globals_0.16.2         goftest_1.2-3          spam_2.10-0           </span></span>
<span><span class="co">#&gt;  [49] rlang_1.1.2            systemfonts_1.0.5      GlobalOptions_0.1.2   </span></span>
<span><span class="co">#&gt;  [52] splines_4.2.1          rstatix_0.7.2          lazyeval_0.2.2        </span></span>
<span><span class="co">#&gt;  [55] spatstat.geom_3.2-7    broom_1.0.5            yaml_2.3.7            </span></span>
<span><span class="co">#&gt;  [58] reshape2_1.4.4         abind_1.4-5            backports_1.4.1       </span></span>
<span><span class="co">#&gt;  [61] httpuv_1.6.12          tools_4.2.1            ggplot2_3.4.4         </span></span>
<span><span class="co">#&gt;  [64] ellipsis_0.3.2         jquerylib_0.1.4        RColorBrewer_1.1-3    </span></span>
<span><span class="co">#&gt;  [67] BiocGenerics_0.44.0    ggridges_0.5.4         Rcpp_1.0.11           </span></span>
<span><span class="co">#&gt;  [70] progress_1.2.2         zlibbioc_1.44.0        purrr_1.0.2           </span></span>
<span><span class="co">#&gt;  [73] RCurl_1.98-1.13        prettyunits_1.2.0      deldir_1.0-9          </span></span>
<span><span class="co">#&gt;  [76] ggpubr_0.6.0           pbapply_1.7-2          GetoptLong_1.0.5      </span></span>
<span><span class="co">#&gt;  [79] cowplot_1.1.1          S4Vectors_0.36.2       zoo_1.8-12            </span></span>
<span><span class="co">#&gt;  [82] ggrepel_0.9.4          cluster_2.1.4          fs_1.6.3              </span></span>
<span><span class="co">#&gt;  [85] magrittr_2.0.3         magick_2.8.1           data.table_1.14.8     </span></span>
<span><span class="co">#&gt;  [88] RSpectra_0.16-1        scattermore_1.2        lmtest_0.9-40         </span></span>
<span><span class="co">#&gt;  [91] RANN_2.6.1             fitdistrplus_1.1-11    matrixStats_1.1.0     </span></span>
<span><span class="co">#&gt;  [94] hms_1.1.3              patchwork_1.1.3        mime_0.12             </span></span>
<span><span class="co">#&gt;  [97] evaluate_0.23          xtable_1.8-4           XML_3.99-0.16         </span></span>
<span><span class="co">#&gt; [100] fastDummies_1.7.3      IRanges_2.32.0         gridExtra_2.3         </span></span>
<span><span class="co">#&gt; [103] shape_1.4.6            compiler_4.2.1         biomaRt_2.54.1        </span></span>
<span><span class="co">#&gt; [106] tibble_3.2.1           KernSmooth_2.23-22     crayon_1.5.2          </span></span>
<span><span class="co">#&gt; [109] htmltools_0.5.7        later_1.3.1            tidyr_1.3.0           </span></span>
<span><span class="co">#&gt; [112] DBI_1.1.3              formatR_1.14           dbplyr_2.4.0          </span></span>
<span><span class="co">#&gt; [115] MASS_7.3-60            rappdirs_0.3.3         Matrix_1.6-3          </span></span>
<span><span class="co">#&gt; [118] car_3.1-2              cli_3.6.1              parallel_4.2.1        </span></span>
<span><span class="co">#&gt; [121] dotCall64_1.1-0        igraph_1.5.1           pkgconfig_2.0.3       </span></span>
<span><span class="co">#&gt; [124] pkgdown_2.0.7          plotly_4.10.3          spatstat.sparse_3.0-3 </span></span>
<span><span class="co">#&gt; [127] xml2_1.3.5             foreach_1.5.2          bslib_0.5.1           </span></span>
<span><span class="co">#&gt; [130] XVector_0.38.0         stringr_1.5.1          digest_0.6.33         </span></span>
<span><span class="co">#&gt; [133] sctransform_0.4.1      RcppAnnoy_0.0.21       spatstat.data_3.0-3   </span></span>
<span><span class="co">#&gt; [136] Biostrings_2.66.0      rmarkdown_2.25         leiden_0.4.3.1        </span></span>
<span><span class="co">#&gt; [139] uwot_0.1.16            curl_5.1.0             shiny_1.8.0           </span></span>
<span><span class="co">#&gt; [142] rjson_0.2.21           nlme_3.1-163           lifecycle_1.0.4       </span></span>
<span><span class="co">#&gt; [145] jsonlite_1.8.7         carData_3.0-5          desc_1.4.2            </span></span>
<span><span class="co">#&gt; [148] viridisLite_0.4.2      fansi_1.0.5            pillar_1.9.0          </span></span>
<span><span class="co">#&gt; [151] lattice_0.22-5         KEGGREST_1.38.0        fastmap_1.1.1         </span></span>
<span><span class="co">#&gt; [154] httr_1.4.7             survival_3.5-7         glue_1.6.2            </span></span>
<span><span class="co">#&gt; [157] png_0.1-8              iterators_1.0.14       bit_4.0.5             </span></span>
<span><span class="co">#&gt; [160] stringi_1.8.2          sass_0.4.7             blob_1.2.4            </span></span>
<span><span class="co">#&gt; [163] textshaping_0.3.7      RcppHNSW_0.5.0         memoise_2.0.1         </span></span>
<span><span class="co">#&gt; [166] dplyr_1.1.4            irlba_2.3.5.1          future.apply_1.11.0</span></span></code></pre></div>
</details>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>domino2 Developed by Christopher Cherry, Jacob T Mitchell, Sushma Nagaraj, Kavita Krishnan.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
