<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="dominoSignal">
<title>SCENIC for TF Activation Scoring • dominoSignal</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Fira_Code-0.4.8/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<script src="../extra.js"></script><meta property="og:title" content="SCENIC for TF Activation Scoring">
<meta property="og:description" content="dominoSignal">
<meta property="og:image" content="https://FertigLab.github.io/dominoSignal/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-#7DDED2"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">dominoSignal</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/dominoSignal.html">Get started</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/cellphonedb_vignette.html">Using the cellphoneDB Database</a>
    <a class="dropdown-item" href="../articles/tf_scenic_vignette.html">SCENIC for TF Activation Scoring</a>
    <a class="dropdown-item" href="../articles/domino_object_vignette.html">Interacting with dominoSignal Objects</a>
    <a class="dropdown-item" href="../articles/plotting_vignette.html">Plotting Functions and Options</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-news">News</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-news">
    <h6 class="dropdown-header" data-toc-skip>Releases</h6>
    <a class="external-link dropdown-item" href="https://github.com/FertigLab/dominoSignal/releases/tag/v0.2.1-alpha">dominoSignal v0.2.1</a>
    <a class="external-link dropdown-item" href="https://github.com/FertigLab/dominoSignal/releases/tag/v0.2.2-alpha">dominoSignal v0.2.2</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../news/index.html">Changelog</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/FertigLab/dominoSignal/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="tf_scenic_vignette_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>SCENIC for TF Activation Scoring</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/FertigLab/dominoSignal/blob/HEAD/vignettes/articles/tf_scenic_vignette.Rmd" class="external-link"><code>vignettes/articles/tf_scenic_vignette.Rmd</code></a></small>
      <div class="d-none name"><code>tf_scenic_vignette.Rmd</code></div>
    </div>

    
    
<p>SCENIC is our lab’s preferred software for TF activity scoring. We recommend using the Python implementation (<a href="https://pyscenic.readthedocs.io/en/latest/" class="external-link">pySCENIC</a>) as it is faster than the original R implementation. Be aware, the use of <a href="https://scenic.aertslab.org/" class="external-link">SCENIC</a> is often the slowest and most memory intensive step of this analysis pipeline. <a href="https://scenic.aertslab.org/" class="external-link">SCENIC</a> should be run on computing resources with access to multi-core processing and large amounts of memory.</p>
<div class="section level2">
<h2 id="file-downloads-for-scenic">File Downloads for SCENIC<a class="anchor" aria-label="anchor" href="#file-downloads-for-scenic"></a>
</h2>
<p>For this tutorial, <a href="https://pyscenic.readthedocs.io/en/latest/" class="external-link">pySCENIC</a> is used as the method for TF activity inference. This requires downloading some files prior to use. Here, we show how to download these files and the way we run pySCENIC to generate the necessary files for use in dominoSignal analyis.</p>
<p>A singularity image of <a href="https://scenic.aertslab.org/" class="external-link">SCENIC</a> v0.12.1 can be installed from the DockerHub image as a source. <a href="https://scenic.aertslab.org/" class="external-link">SCENIC</a> requires a list of TFs, motif annotations, and cisTarget motifs which are all available from the authors of <a href="https://scenic.aertslab.org/" class="external-link">SCENIC</a> for human (HGNC), mouse (MGI), and fly. The following will download everything necessary for an analysis of a data set with HGNC gene labels for the hg38 genome.</p>
<div class="section level3">
<h3 id="reference-files">Reference files<a class="anchor" aria-label="anchor" href="#reference-files"></a>
</h3>
<p>First we set up a temporary directory to store the files.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">temp_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Then we build the singularity image and download reference files to the temporary directory.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1"></a><span class="va">SCENIC_DIR=</span><span class="st">"'temp_dir'/scenic"</span></span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="fu">mkdir</span> <span class="va">${SCENIC_DIR}</span></span>
<span id="cb2-3"><a href="#cb2-3"></a></span>
<span id="cb2-4"><a href="#cb2-4"></a><span class="co"># Build singularity image</span></span>
<span id="cb2-5"><a href="#cb2-5"></a><span class="ex">singularity</span> build <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/aertslab-pyscenic-0.12.1.sif"</span> docker://aertslab/pyscenic:0.12.1</span>
<span id="cb2-6"><a href="#cb2-6"></a></span>
<span id="cb2-7"><a href="#cb2-7"></a><span class="co"># Matrix containing motifs as rows and genes as columns and ranking position for each gene and motif (based on CRM scores) as values. URLs provided link to v2 feather files required for the 0.12.1 version of pySENIC.</span></span>
<span id="cb2-8"><a href="#cb2-8"></a></span>
<span id="cb2-9"><a href="#cb2-9"></a><span class="ex">curl</span> <span class="st">"https://resources.aertslab.org/cistarget/databases/homo_sapiens/hg38/refseq_r80/mc_v10_clust/gene_based/hg38_10kbp_up_10kbp_down_full_tx_v10_clust.genes_vs_motifs.rankings.feather"</span> \</span>
<span id="cb2-10"><a href="#cb2-10"></a>  -o <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/hg38_10kbp_up_10kbp_down_full_tx_v10_clust.genes_vs_motifs.rankings.feather"</span></span>
<span id="cb2-11"><a href="#cb2-11"></a></span>
<span id="cb2-12"><a href="#cb2-12"></a><span class="ex">curl</span> <span class="st">"https://resources.aertslab.org/cistarget/databases/homo_sapiens/hg38/refseq_r80/mc_v10_clust/gene_based/hg38_500bp_up_100bp_down_full_tx_v10_clust.genes_vs_motifs.rankings.feather"</span> \</span>
<span id="cb2-13"><a href="#cb2-13"></a>  -o <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/hg38_500bp_up_100bp_down_full_tx_v10_clust.genes_vs_motifs.rankings.feather"</span></span>
<span id="cb2-14"><a href="#cb2-14"></a></span>
<span id="cb2-15"><a href="#cb2-15"></a><span class="co"># List of genes encoding TFs in the HG38 reference genome</span></span>
<span id="cb2-16"><a href="#cb2-16"></a><span class="ex">curl</span> <span class="st">"https://resources.aertslab.org/cistarget/tf_lists/allTFs_hg38.txt"</span> \</span>
<span id="cb2-17"><a href="#cb2-17"></a>  -o <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/allTFs_hg38.txt"</span></span>
<span id="cb2-18"><a href="#cb2-18"></a></span>
<span id="cb2-19"><a href="#cb2-19"></a><span class="co"># Motif annotations based on the 2017 cisTarget motif collection. Use these files if you are using the mc9nr databases.</span></span>
<span id="cb2-20"><a href="#cb2-20"></a><span class="ex">curl</span> <span class="st">"https://resources.aertslab.org/cistarget/motif2tf/motifs-v9-nr.hgnc-m0.001-o0.0.tbl"</span> \</span>
<span id="cb2-21"><a href="#cb2-21"></a>  -o <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/motifs-v9-nr.hgnc-m0.001-o0.0.tbl"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="preparing-preprocessed-data">Preparing Preprocessed Data:<a class="anchor" aria-label="anchor" href="#preparing-preprocessed-data"></a>
</h3>
<p>dominoSignal was designed to be compatible with single cell workflows, accepting gene by cell matrices for counts and scaled counts. However, <a href="https://pyscenic.readthedocs.io/en/latest/" class="external-link">pySCENIC</a> is used for TF activation inference, and it requires as input a cell by gene matrix. This is the opposite orientation of many R based single cell packages (such as <a href="https://satijalab.org/seurat" class="external-link">Seurat</a> or <code>{SingleCellExperiment}</code>) but is the default for Python based tools such as <a href="https://scanpy.readthedocs.io/en/stable/" class="external-link">scanpy</a>. The RNA counts matrix can be saved as a tab-seperated value (.tsv) or comma-sperated value (.csv) file after transposing the matrix to cell by gene orientation. Below is an example of extracting the counts from a <code>{SingleCellExperiment}</code> object and saving it as a .tsv file.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_counts</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"counts"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">pbmc_counts</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">input_dir</span>, <span class="st">"/pbmc3k_counts.tsv"</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"\t"</span>,</span>
<span>    col.names <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span></code></pre></div>
<p>tsv and csv files are very inefficient for storing large matrices. As an alternative, the counts matrix can be saved as a loom file and directly passed to <a href="https://pyscenic.readthedocs.io/en/latest/" class="external-link">pySCENIC</a> in this format. Generating a loom file in R requires use of the <a href="https://github.com/mojaveazure/loomR" class="external-link">loomR</a> package. The package automatically handles the transposition of the counts matrix to cell by gene orientation as well. We recommend this approach to passing a counts matrix to pySCENIC. However, users should be aware that <a href="https://github.com/mojaveazure/loomR" class="external-link">loomR</a> is not hosted on CRAN or BioConductor at the time of this vignette’s creation.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mojaveazure/loomR" class="external-link">loomR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># save loom counts matrix</span></span>
<span><span class="va">pbmc_counts</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"counts"</span><span class="op">)</span></span>
<span><span class="va">pbmc_loom</span> <span class="op">&lt;-</span> <span class="fu">loomR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/loomR/man/create.html" class="external-link">create</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">input_dir</span>, <span class="st">"/pbmc3k_counts.loom"</span><span class="op">)</span>, data <span class="op">=</span> <span class="va">pbmc_counts</span><span class="op">)</span></span>
<span><span class="co"># connection to the loom file must be closed to complete file creation</span></span>
<span><span class="va">pbmc_loom</span><span class="op">$</span><span class="fu">close_all</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>We will use the <code>pbmc3k\_counts.loom</code> file in this tutorial.</p>
</div>
</div>
<div class="section level2">
<h2 id="running-scenic">Running SCENIC<a class="anchor" aria-label="anchor" href="#running-scenic"></a>
</h2>
<p><a href="https://pyscenic.readthedocs.io/en/latest/" class="external-link">pySCENIC</a> is initiated using bash scripting in the terminal. The analysis consists of 3 steps to score genes for TF motif enrichment, construct TF regulons consisting of genes targeted by the TFs, and arrive at AUC scores for enrichment of regulon gene transcription within each cell.</p>
<div class="section level3">
<h3 id="grn-construct-tf-modules">grn: construct TF-modules<a class="anchor" aria-label="anchor" href="#grn-construct-tf-modules"></a>
</h3>
<p>Co-expression modules are used to quantify gene-TF adjacencies.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1"></a><span class="ex">singularity</span> exec <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/aertslab-pyscenic-0.12.1.sif"</span> pyscenic grn \</span>
<span id="cb5-2"><a href="#cb5-2"></a>    <span class="st">"pbmc3k_counts.loom"</span> \</span>
<span id="cb5-3"><a href="#cb5-3"></a>    <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/allTFs_hg38.txt"</span> \</span>
<span id="cb5-4"><a href="#cb5-4"></a>    -o <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/pbmc_adj_3k.tsv"</span> \</span>
<span id="cb5-5"><a href="#cb5-5"></a>    --num_workers 6 \</span>
<span id="cb5-6"><a href="#cb5-6"></a>    --seed 123</span>
<span id="cb5-7"><a href="#cb5-7"></a></span>
<span id="cb5-8"><a href="#cb5-8"></a><span class="co"># Arguments:</span></span>
<span id="cb5-9"><a href="#cb5-9"></a>    <span class="co"># path to the loom file</span></span>
<span id="cb5-10"><a href="#cb5-10"></a>    <span class="co"># list of TFs</span></span>
<span id="cb5-11"><a href="#cb5-11"></a>    <span class="co"># output directory for the adjacency matrix</span></span>
<span id="cb5-12"><a href="#cb5-12"></a>    <span class="co"># number of CPUs to use if multi-core processing is available</span></span>
<span id="cb5-13"><a href="#cb5-13"></a>    <span class="co"># specify a random seed for reproducibility</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="ctx-construct-tf-regulons-with-pruning-based-on-tf-motif-enrichment">ctx: construct TF regulons with pruning based on TF motif enrichment<a class="anchor" aria-label="anchor" href="#ctx-construct-tf-regulons-with-pruning-based-on-tf-motif-enrichment"></a>
</h3>
<p>The rankings of genes based on enrichment of TF motifs to the transcription start site (TSS) are considered in the construction of regulons, where target genes in the TF-modules are removed if they lack motifs proximal to the TSS where the TF may bind.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1"></a><span class="ex">singularity</span> exec <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/aertslab-pyscenic-0.12.1.sif"</span> pyscenic ctx \</span>
<span id="cb6-2"><a href="#cb6-2"></a>    <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/pbmc_adj.tsv"</span> \</span>
<span id="cb6-3"><a href="#cb6-3"></a>    <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/hg38_10kbp_up_10kbp_down_full_tx_v10_clust.genes_vs_motifs.rankings.feather"</span> \</span>
<span id="cb6-4"><a href="#cb6-4"></a>    <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/hg38_500bp_up_100bp_down_full_tx_v10_clust.genes_vs_motifs.rankings.feather"</span> \</span>
<span id="cb6-5"><a href="#cb6-5"></a>    --annotations_fname <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/motifs-v9-nr.hgnc-m0.001-o0.0.tbl"</span> \</span>
<span id="cb6-6"><a href="#cb6-6"></a>    --expression_mtx_fname <span class="st">"pbmc3k_counts.loom"</span> \</span>
<span id="cb6-7"><a href="#cb6-7"></a>    --mode <span class="st">"dask_multiprocessing"</span> \</span>
<span id="cb6-8"><a href="#cb6-8"></a>    --output <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/regulons_pbmc_3k.csv"</span> \</span>
<span id="cb6-9"><a href="#cb6-9"></a>    --num_workers 1                                                          </span>
<span id="cb6-10"><a href="#cb6-10"></a></span>
<span id="cb6-11"><a href="#cb6-11"></a><span class="co"># Arguments:</span></span>
<span id="cb6-12"><a href="#cb6-12"></a>    <span class="co"># adjacency matrix output from grn</span></span>
<span id="cb6-13"><a href="#cb6-13"></a>    <span class="co"># target rankings of motif enrichment within 10 kb of TSS</span></span>
<span id="cb6-14"><a href="#cb6-14"></a>    <span class="co"># target rankings of motif enrichment within 500 bp upstream and 100 bp downstream of TSS</span></span>
<span id="cb6-15"><a href="#cb6-15"></a>    <span class="co"># TF motif features</span></span>
<span id="cb6-16"><a href="#cb6-16"></a>    <span class="co"># counts matrix loom file</span></span>
<span id="cb6-17"><a href="#cb6-17"></a>    <span class="co"># enable multi-core processing</span></span>
<span id="cb6-18"><a href="#cb6-18"></a>    <span class="co"># output file of learned TF regulons</span></span>
<span id="cb6-19"><a href="#cb6-19"></a>    <span class="co"># number of CPU cores</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="aucell-calculate-tf-activity-scores">aucell: calculate TF activity scores<a class="anchor" aria-label="anchor" href="#aucell-calculate-tf-activity-scores"></a>
</h3>
<p>Enrichment of a regulon is measured as the <strong>A</strong>rea <strong>U</strong>nder the recovery <strong>C</strong>urve (AUC) of the genes that define this regulon.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1"></a><span class="ex">singularity</span> exec <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/aertslab-pyscenic-0.12.1.sif"</span> pyscenic aucell \</span>
<span id="cb7-2"><a href="#cb7-2"></a>    <span class="st">"pbmc3k_counts.loom"</span> \</span>
<span id="cb7-3"><a href="#cb7-3"></a>    <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/regulons_pbmc_3k.csv"</span> \</span>
<span id="cb7-4"><a href="#cb7-4"></a>    -o <span class="st">"</span><span class="va">${SCENIC_DIR}</span><span class="st">/auc_pbmc_3k.csv"</span></span>
<span id="cb7-5"><a href="#cb7-5"></a>    </span>
<span id="cb7-6"><a href="#cb7-6"></a><span class="co"># Arguments:</span></span>
<span id="cb7-7"><a href="#cb7-7"></a>    <span class="co"># counts matrix loom file</span></span>
<span id="cb7-8"><a href="#cb7-8"></a>    <span class="co"># regulon table output from ctx</span></span>
<span id="cb7-9"><a href="#cb7-9"></a>    <span class="co"># cell x TF matrix of TF enrichment AUC values</span></span></code></pre></div>
<p>The csv files created in this analysis can be used as the TF activation inputs in a dominoSignal analysis. Please see our <a href="vignette(%22dominoSignal%22)">Getting Started</a> page for more information on how to perform an analysis with dominoSignal. For a brief tutorial on the download of the necessary files for using cellphonedb, please see the <a href="vignette(%22cellphonedb_vignette%22)">Using the cellphoneDB database</a> vignette.</p>
<details><p><summary>Vignette Build Information</summary> Date last built and session information:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.Date</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "2024-03-27"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0</span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] digest_0.6.35     R6_2.5.1          lifecycle_1.0.4   jsonlite_1.8.8   </span></span>
<span><span class="co">#&gt;  [5] formatR_1.14      magrittr_2.0.3    evaluate_0.23     rlang_1.1.3      </span></span>
<span><span class="co">#&gt;  [9] cachem_1.0.8      cli_3.6.2         fs_1.6.3          jquerylib_0.1.4  </span></span>
<span><span class="co">#&gt; [13] bslib_0.6.1       ragg_1.3.0        vctrs_0.6.5       rmarkdown_2.26   </span></span>
<span><span class="co">#&gt; [17] pkgdown_2.0.7     textshaping_0.3.7 desc_1.4.3        tools_4.2.1      </span></span>
<span><span class="co">#&gt; [21] purrr_1.0.2       yaml_2.3.8        xfun_0.42         fastmap_1.1.1    </span></span>
<span><span class="co">#&gt; [25] compiler_4.2.1    systemfonts_1.0.6 memoise_2.0.1     htmltools_0.5.7  </span></span>
<span><span class="co">#&gt; [29] knitr_1.45        sass_0.4.8</span></span></code></pre></div>
</details>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>dominoSignal is developed by Christopher Cherry, Jacob T Mitchell, Sushma Nagaraj, Kavita Krishnan, Dmitrijs Lvovs.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
