<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function reads in a receptor ligand signaling database, cell level
features of some kind (ie. output from pySCENIC), z-scored single cell data,
and cluster id for single cell data, calculates a correlation matrix between
receptors and other features (this is transcription factor module scores if
using pySCENIC), and finds features enriched by cluster. It will return a
domino object prepared for build_domino(), which will calculate a signaling
network."><title>Create a domino object and prepare it for network construction — create_domino • domino2</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><script src="../extra.js"></script><meta property="og:title" content="Create a domino object and prepare it for network construction — create_domino"><meta property="og:description" content="This function reads in a receptor ligand signaling database, cell level
features of some kind (ie. output from pySCENIC), z-scored single cell data,
and cluster id for single cell data, calculates a correlation matrix between
receptors and other features (this is transcription factor module scores if
using pySCENIC), and finds features enriched by cluster. It will return a
domino object prepared for build_domino(), which will calculate a signaling
network."><meta property="og:image" content="/logo.svg"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-#7DDED2"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">domino2</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/domino2.html">Get started</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/domino_object_vignette.html">Interacting with domino2 Objects</a>
    <a class="dropdown-item" href="../articles/plotting_vignette.html">Plotting Functions and Options</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/FertigLab/domino_development/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Create a domino object and prepare it for network construction</h1>
      <small class="dont-index">Source: <a href="https://github.com/FertigLab/domino_development/blob/HEAD/R/import_fxns.R" class="external-link"><code>R/import_fxns.R</code></a></small>
      <div class="d-none name"><code>create_domino.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function reads in a receptor ligand signaling database, cell level
features of some kind (ie. output from pySCENIC), z-scored single cell data,
and cluster id for single cell data, calculates a correlation matrix between
receptors and other features (this is transcription factor module scores if
using pySCENIC), and finds features enriched by cluster. It will return a
domino object prepared for <code><a href="build_domino.html">build_domino()</a></code>, which will calculate a signaling
network.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">create_domino</span><span class="op">(</span></span>
<span>  <span class="va">rl_map</span>,</span>
<span>  <span class="va">features</span>,</span>
<span>  ser <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  counts <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  z_scores <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  clusters <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  use_clusters <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  tf_targets <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  use_complexes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  rec_min_thresh <span class="op">=</span> <span class="fl">0.025</span>,</span>
<span>  remove_rec_dropout <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  tf_selection_method <span class="op">=</span> <span class="st">"clusters"</span>,</span>
<span>  tf_variance_quantile <span class="op">=</span> <span class="fl">0.5</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>rl_map</dt>
<dd><p>Data frame where each row describes a receptor-ligand interaction with required columns gene_A &amp; gene_B including the gene names for the receptor and ligand and type_A &amp; type_B annotating if genes A and B are a ligand (L) or receptor (R)</p></dd>


<dt>features</dt>
<dd><p>Either a path to a csv containing cell level features of interest (ie. the auc matrix from pySCENIC) or named matrix with cells as columns and features as rows.</p></dd>


<dt>ser</dt>
<dd><p>Seurat object containing scaled RNA expression data in the RNA assay slot and cluster identity. Either a ser object OR z_scores and clusters must be provided. If ser is present z_scores and clusters will be ignored.</p></dd>


<dt>counts</dt>
<dd><p>Counts matrix for the data. If a Seurat object is provided this will be ignored. This is only used to threshold receptors on dropout.</p></dd>


<dt>z_scores</dt>
<dd><p>Matrix containing z-scored expression data for all cells with cells as columns and features as rows. Either z_scores and clusters must be provided OR a ser object. If ser is present z_scores and clusters will be ignored.</p></dd>


<dt>clusters</dt>
<dd><p>Named factor containing cell cluster with names as cells. Either clusters and z_scores OR ser must be provided. If ser is present z_scores and clusters will be ignored.</p></dd>


<dt>use_clusters</dt>
<dd><p>Boolean indicating whether to use the clusters from a Seurat object. If a Seurat object is not provided then this parameter is ignored.</p></dd>


<dt>tf_targets</dt>
<dd><p>Optional. A list where names are transcription factors and the stored values are character vectors of genes in the transcription factor's regulon.</p></dd>


<dt>verbose</dt>
<dd><p>Boolean indicating whether or not to print progress during computation.</p></dd>


<dt>use_complexes</dt>
<dd><p>Boolean indicating whether you wish to use receptor/ligand complexes in the receptor ligand signaling database. If FALSE, receptor/ligand pairs where either functions as a protein complex will not be considered when constructing the signaling network.</p></dd>


<dt>rec_min_thresh</dt>
<dd><p>Minimum expression level of receptors by cell. Default is 0.025 or 2.5 percent of all cells in the data set. This is important when calculating correlation to connect receptors to transcription activation. If this threshold is too low then correlation calculations will proceed with very few cells with non-zero expression.</p></dd>


<dt>remove_rec_dropout</dt>
<dd><p>Whether to remove receptors with 0 expression counts when calculating correlations. This can reduce false positive correlation calculations when receptors have high dropout rates.</p></dd>


<dt>tf_selection_method</dt>
<dd><p>Selection of which method to target transcription factors. If 'clusters' then differential expression for clusters will be calculated. If 'variable' then the most variable transcription factors will be selected. If 'all' then all transcription factors in the feature matrix will be used. Default is 'clusters'. Note that if you wish to use clusters for intercellular signaling downstream to MUST choose clusters.</p></dd>


<dt>tf_variance_quantile</dt>
<dd><p>What proportion of variable features to take if using variance to threshold features. Default is 0.5. Higher numbers will keep more features. Ignored if tf_selection_method is not 'variable'</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A domino object</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">pbmc_dom_tiny_all</span> <span class="op">&lt;-</span> <span class="va">pbmc_dom_tiny</span> <span class="op">&lt;-</span> <span class="fu">create_domino</span><span class="op">(</span></span></span>
<span class="r-in"><span>    rl_map <span class="op">=</span> <span class="fu">domino2</span><span class="fu">:::</span><span class="va">rl_map_tiny</span>, features <span class="op">=</span> <span class="fu">domino2</span><span class="fu">:::</span><span class="va">auc_tiny</span>,</span></span>
<span class="r-in"><span>    counts <span class="op">=</span> <span class="fu">domino2</span><span class="fu">:::</span><span class="va">RNA_count_tiny</span>, z_scores <span class="op">=</span> <span class="fu">domino2</span><span class="fu">:::</span><span class="va">RNA_zscore_tiny</span>,</span></span>
<span class="r-in"><span>    clusters <span class="op">=</span> <span class="fu">domino2</span><span class="fu">:::</span><span class="va">clusters_tiny</span>, tf_targets <span class="op">=</span> <span class="fu">domino2</span><span class="fu">:::</span><span class="va">regulon_list_tiny</span>,</span></span>
<span class="r-in"><span>    use_clusters <span class="op">=</span> <span class="cn">FALSE</span>, use_complexes <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>    rec_min_thresh <span class="op">=</span> <span class="fl">0.1</span>, remove_rec_dropout <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>    tf_selection_method <span class="op">=</span> <span class="st">"all"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading in and processing signaling database</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Database provided from source: CellPhoneDB</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Getting z_scores, clusters, and counts</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating correlations</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 1 of 3</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2 of 3</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 3 of 3</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> FAS has component genes that did not pass testing parameters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> CD22 has component genes that did not pass testing parameters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> IGF1R has component genes that did not pass testing parameters</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">pbmc_dom_tiny_clustered</span> <span class="op">&lt;-</span> <span class="fu">create_domino</span><span class="op">(</span></span></span>
<span class="r-in"><span>    rl_map <span class="op">=</span> <span class="fu">domino2</span><span class="fu">:::</span><span class="va">rl_map_tiny</span>, features <span class="op">=</span> <span class="fu">domino2</span><span class="fu">:::</span><span class="va">auc_tiny</span>,</span></span>
<span class="r-in"><span>    counts <span class="op">=</span> <span class="fu">domino2</span><span class="fu">:::</span><span class="va">RNA_count_tiny</span>, z_scores <span class="op">=</span> <span class="fu">domino2</span><span class="fu">:::</span><span class="va">RNA_zscore_tiny</span>,</span></span>
<span class="r-in"><span>    clusters <span class="op">=</span> <span class="fu">domino2</span><span class="fu">:::</span><span class="va">clusters_tiny</span>, tf_targets <span class="op">=</span> <span class="fu">domino2</span><span class="fu">:::</span><span class="va">regulon_list_tiny</span>,</span></span>
<span class="r-in"><span>    use_clusters <span class="op">=</span> <span class="cn">TRUE</span>, use_complexes <span class="op">=</span> <span class="cn">TRUE</span>, remove_rec_dropout <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading in and processing signaling database</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Database provided from source: CellPhoneDB</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Getting z_scores, clusters, and counts</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating feature enrichment by cluster</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 1 of 3</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2 of 3</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 3 of 3</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating correlations</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 1 of 3</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Cannot compute exact p-value with ties</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Cannot compute exact p-value with ties</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Cannot compute exact p-value with ties</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Cannot compute exact p-value with ties</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Cannot compute exact p-value with ties</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2 of 3</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Cannot compute exact p-value with ties</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Cannot compute exact p-value with ties</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Cannot compute exact p-value with ties</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Cannot compute exact p-value with ties</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Cannot compute exact p-value with ties</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 3 of 3</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Cannot compute exact p-value with ties</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Cannot compute exact p-value with ties</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Cannot compute exact p-value with ties</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Cannot compute exact p-value with ties</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Cannot compute exact p-value with ties</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> IGF1R has component genes that did not pass testing parameters</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>domino2 Developed by Christopher Cherry, Jacob T Mitchell, Sushma Nagaraj, Kavita Krishnan.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

