<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="domino2">
<title>Interacting with domino2 Objects • domino2</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<script src="../extra.js"></script><meta property="og:title" content="Interacting with domino2 Objects">
<meta property="og:description" content="domino2">
<meta property="og:image" content="https://FertigLab.github.io/domino2/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-#7DDED2"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">domino2</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/domino2.html">Get started</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/domino_object_vignette.html">Interacting with domino2 Objects</a>
    <a class="dropdown-item" href="../articles/plotting_vignette.html">Plotting Functions and Options</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-news">News</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-news">
    <h6 class="dropdown-header" data-toc-skip>Releases</h6>
    <a class="external-link dropdown-item" href="https://github.com/FertigLab/domino2/releases/tag/v0.2.1-alpha">domino2 v0.2.1</a>
    <a class="external-link dropdown-item" href="https://github.com/FertigLab/domino2/releases/tag/v0.2.2-alpha">domino2 v0.2.2</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../news/index.html">Changelog</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/FertigLab/domino2/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Interacting with domino2 Objects</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/FertigLab/domino2/blob/HEAD/vignettes/domino_object_vignette.Rmd" class="external-link"><code>vignettes/domino_object_vignette.Rmd</code></a></small>
      <div class="d-none name"><code>domino_object_vignette.Rmd</code></div>
    </div>

    
    
<p>In order to better implement some of the changes that have been made
in domino2, the structure of the domino object has been modified. Here,
we describe the data that is stored in the domino object, as well as how
best to access this data.</p>
<div class="section level2">
<h2 id="object-contents">Object contents<a class="anchor" aria-label="anchor" href="#object-contents"></a>
</h2>
<p>The contents of the <a href="%60domino-class%60">domino object
class</a> include a few broad groups: - input data, such as the L-R
database used, counts, and z-scores - calculations, such as correlations
or differential expression results - linkages, which show connections
between TFs, receptors, and ligands across the data set and within
clusters - signaling matrices, which show expression of ligands
signaling to active receptors in clusters (or a summary of overall
signaling) - build information, which includes the parameters used to
build the object in the <code><a href="../reference/build_domino.html">build_domino()</a></code> functions</p>
<p>For commonly accessed information (the number of cells, clusters, and
some build information), the show and print methods for domino objects
can be used.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dom</span></span>
<span><span class="co">#&gt; A domino object of 2700 cells</span></span>
<span><span class="co">#&gt; Built with signaling between 9 clusters</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">dom</span><span class="op">)</span></span>
<span><span class="co">#&gt; A domino object of 2700 cells</span></span>
<span><span class="co">#&gt; Built with signaling between 9 clusters</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="access-functions">Access functions<a class="anchor" aria-label="anchor" href="#access-functions"></a>
</h2>
<p>In order to facilitate access to the information stored in the domino
object, we have provided a collection of functions to retrieve specific
items. These functions begin with “dom_” and can be listed using
<code><a href="https://rdrr.io/r/base/ls.html" class="external-link">ls()</a></code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/ls.html" class="external-link">ls</a></span><span class="op">(</span><span class="st">"package:domino2"</span>, pattern <span class="op">=</span> <span class="st">"^dom_"</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "dom_clusters"      "dom_correlations"  "dom_counts"       </span></span>
<span><span class="co">#&gt;  [4] "dom_database"      "dom_de"            "dom_info"         </span></span>
<span><span class="co">#&gt;  [7] "dom_linkages"      "dom_network_items" "dom_signaling"    </span></span>
<span><span class="co">#&gt; [10] "dom_tf_activation" "dom_zscores"</span></span></code></pre></div>
<div class="section level3">
<h3 id="input-data">Input data<a class="anchor" aria-label="anchor" href="#input-data"></a>
</h3>
<p>When creating a domino object with the <code><a href="../reference/create_domino.html">create_domino()</a></code>
function, several inputs are required which are then stored in the
domino object itself. These include cluster labels, the counts matrix,
z-scored counts matrix, transcription factor activation scores, and the
R-L database used in <code><a href="../reference/create_rl_map_cellphonedb.html">create_rl_map_cellphonedb()</a></code>.</p>
<p>For example, to access the cluster names in the domino object:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/dom_clusters.html">dom_clusters</a></span><span class="op">(</span><span class="va">dom</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "naive_CD4_T_cell"  "CD14_monocyte"     "memory_CD4_T_cell"</span></span>
<span><span class="co">#&gt; [4] "B_cell"            "CD8_T_cell"        "CD16_monocyte"    </span></span>
<span><span class="co">#&gt; [7] "NK_cell"           "dendritic_cell"    "platelet"</span></span></code></pre></div>
<p>Setting an argument <code>labels = TRUE</code> will return the vector
of cluster labels for each cell rather than the unique cluster
names.</p>
<p>To access the counts:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">count_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dom_counts.html">dom_counts</a></span><span class="op">(</span><span class="va">dom</span><span class="op">)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">count_matrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="14%">
<col width="17%">
<col width="17%">
<col width="17%">
<col width="17%">
<col width="17%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="right">AAACATACAACCAC-1</th>
<th align="right">AAACATTGAGCTAC-1</th>
<th align="right">AAACATTGATCAGC-1</th>
<th align="right">AAACCGTGCTTCCG-1</th>
<th align="right">AAACCGTGTATGCG-1</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">AL627309.1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">AP006222.2</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">RP11-206L10.2</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">RP11-206L10.9</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">LINC00115</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
<p>Or z-scored counts:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">z_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dom_zscores.html">dom_zscores</a></span><span class="op">(</span><span class="va">dom</span><span class="op">)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">z_matrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="14%">
<col width="17%">
<col width="17%">
<col width="17%">
<col width="17%">
<col width="17%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="right">AAACATACAACCAC-1</th>
<th align="right">AAACATTGAGCTAC-1</th>
<th align="right">AAACATTGATCAGC-1</th>
<th align="right">AAACCGTGCTTCCG-1</th>
<th align="right">AAACCGTGTATGCG-1</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">AL627309.1</td>
<td align="right">-0.0574500</td>
<td align="right">-0.0574500</td>
<td align="right">-0.0574500</td>
<td align="right">-0.0574500</td>
<td align="right">-0.0574500</td>
</tr>
<tr class="even">
<td align="left">AP006222.2</td>
<td align="right">-0.0331877</td>
<td align="right">-0.0331877</td>
<td align="right">-0.0331877</td>
<td align="right">-0.0331877</td>
<td align="right">-0.0331877</td>
</tr>
<tr class="odd">
<td align="left">RP11-206L10.2</td>
<td align="right">-0.0411864</td>
<td align="right">-0.0411864</td>
<td align="right">-0.0411864</td>
<td align="right">-0.0411864</td>
<td align="right">-0.0411864</td>
</tr>
<tr class="even">
<td align="left">RP11-206L10.9</td>
<td align="right">-0.0332568</td>
<td align="right">-0.0332568</td>
<td align="right">-0.0332568</td>
<td align="right">-0.0332568</td>
<td align="right">-0.0332568</td>
</tr>
<tr class="odd">
<td align="left">LINC00115</td>
<td align="right">-0.0812841</td>
<td align="right">-0.0812841</td>
<td align="right">-0.0812841</td>
<td align="right">-0.0812841</td>
<td align="right">-0.0812841</td>
</tr>
</tbody>
</table>
<p>The transcription factor activation scores can be similarly
accessed:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">activation_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dom_tf_activation.html">dom_tf_activation</a></span><span class="op">(</span><span class="va">dom</span><span class="op">)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">activation_matrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<table style="width:100%;" class="table">
<colgroup>
<col width="7%">
<col width="18%">
<col width="18%">
<col width="18%">
<col width="18%">
<col width="18%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="right">AAACATACAACCAC-1</th>
<th align="right">AAACATTGAGCTAC-1</th>
<th align="right">AAACATTGATCAGC-1</th>
<th align="right">AAACCGTGCTTCCG-1</th>
<th align="right">AAACCGTGTATGCG-1</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">ARID3A</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.1825397</td>
<td align="right">0.0000000</td>
</tr>
<tr class="even">
<td align="left">ARNTL</td>
<td align="right">0.0000000</td>
<td align="right">0.0338921</td>
<td align="right">0.0000000</td>
<td align="right">0.0276239</td>
<td align="right">0.0131924</td>
</tr>
<tr class="odd">
<td align="left">ATF1</td>
<td align="right">0.2017909</td>
<td align="right">0.1039150</td>
<td align="right">0.1261974</td>
<td align="right">0.1303623</td>
<td align="right">0.0000000</td>
</tr>
<tr class="even">
<td align="left">ATF2</td>
<td align="right">0.0754373</td>
<td align="right">0.0822400</td>
<td align="right">0.1471088</td>
<td align="right">0.0693635</td>
<td align="right">0.0710641</td>
</tr>
<tr class="odd">
<td align="left">ATF3</td>
<td align="right">0.0678526</td>
<td align="right">0.0763180</td>
<td align="right">0.0774797</td>
<td align="right">0.1279762</td>
<td align="right">0.0527818</td>
</tr>
</tbody>
</table>
<p>Information about the database referenced for ligand-receptor pairs
and composition of protien complexes can be extracted from the
<code><a href="../reference/dom_database.html">dom_database()</a></code> function. By default, the function returns
the name(s) of the database(s) used:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/dom_database.html">dom_database</a></span><span class="op">(</span><span class="va">dom</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "CellPhoneDB_v4.0"</span></span></code></pre></div>
<p>If you would like to view the entire ligand-receptor map, set
<code>name_only = FALSE</code>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">db_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dom_database.html">dom_database</a></span><span class="op">(</span><span class="va">dom</span>, name_only <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">db_matrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">int_pair</th>
<th align="left">name_A</th>
<th align="left">uniprot_A</th>
<th align="left">gene_A</th>
<th align="left">type_A</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">FGF4 &amp; FGFR1</td>
<td align="left">FGF4</td>
<td align="left">P08620</td>
<td align="left">FGF4</td>
<td align="left">L</td>
</tr>
<tr class="even">
<td align="left">ADORA3 &amp; ENTPD1</td>
<td align="left">ADORA3</td>
<td align="left">P0DMS8</td>
<td align="left">ADORA3</td>
<td align="left">R</td>
</tr>
<tr class="odd">
<td align="left">EPHB3 &amp; EFNB1</td>
<td align="left">EPHB3</td>
<td align="left">P54753</td>
<td align="left">EPHB3</td>
<td align="left">R</td>
</tr>
<tr class="even">
<td align="left">TNFSF18 &amp; TNFRSF18</td>
<td align="left">TNFSF18</td>
<td align="left">Q9UNG2</td>
<td align="left">TNFSF18</td>
<td align="left">L</td>
</tr>
<tr class="odd">
<td align="left">EFNA1 &amp; EPHA1</td>
<td align="left">EFNA1</td>
<td align="left">P20827</td>
<td align="left">EFNA1</td>
<td align="left">L</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="calculations">Calculations<a class="anchor" aria-label="anchor" href="#calculations"></a>
</h3>
<p>Active transcription factors in each cluster are determined by
conducting wilcoxon rank sum tests for each transcription factor where
the transcription factor activity scores amongst all cells in a cluster
are tested against the activity scores of all cells outside of the
cluster. The p-values for the one-sided test for greater activity within
the cluster compared to outside can be accessed with the
<code><a href="../reference/dom_de.html">dom_de()</a></code> function.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">de_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dom_de.html">dom_de</a></span><span class="op">(</span><span class="va">dom</span><span class="op">)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">de_matrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="9%">
<col width="22%">
<col width="18%">
<col width="23%">
<col width="12%">
<col width="14%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="right">naive_CD4_T_cell</th>
<th align="right">CD14_monocyte</th>
<th align="right">memory_CD4_T_cell</th>
<th align="right">B_cell</th>
<th align="right">CD8_T_cell</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">ARID3A</td>
<td align="right">1.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.9999909</td>
<td align="right">0.9999718</td>
<td align="right">0.9999031</td>
</tr>
<tr class="even">
<td align="left">ARNTL</td>
<td align="right">1.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.3287915</td>
<td align="right">1.0000000</td>
<td align="right">0.8408834</td>
</tr>
<tr class="odd">
<td align="left">ATF1</td>
<td align="right">0.0942821</td>
<td align="right">0.7919584</td>
<td align="right">0.0000000</td>
<td align="right">0.9999999</td>
<td align="right">0.8473444</td>
</tr>
<tr class="even">
<td align="left">ATF2</td>
<td align="right">0.0000002</td>
<td align="right">1.0000000</td>
<td align="right">0.0000000</td>
<td align="right">1.0000000</td>
<td align="right">0.6222360</td>
</tr>
<tr class="odd">
<td align="left">ATF3</td>
<td align="right">1.0000000</td>
<td align="right">0.0000000</td>
<td align="right">1.0000000</td>
<td align="right">1.0000000</td>
<td align="right">1.0000000</td>
</tr>
</tbody>
</table>
<p>Linkage between receptors and transcription factors is assessed by
Spearman correlation between transcription factor activity scores and
scaled expression of receptor-encoding genes across all cells in the
data set. Spearman coefficients can be accessed with the
<code><a href="../reference/dom_correlations.html">dom_correlations()</a></code> function.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cor_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dom_correlations.html">dom_correlations</a></span><span class="op">(</span><span class="va">dom</span><span class="op">)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">cor_matrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">ARID3A</th>
<th align="right">ARNTL</th>
<th align="right">ATF1</th>
<th align="right">ATF2</th>
<th align="right">ATF3</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">FGFR1</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
</tr>
<tr class="even">
<td align="left">ADORA3</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
</tr>
<tr class="odd">
<td align="left">EPHB3</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
</tr>
<tr class="even">
<td align="left">TNFRSF18</td>
<td align="right">-0.0415026</td>
<td align="right">0.0292465</td>
<td align="right">0.0411616</td>
<td align="right">0.0595802</td>
<td align="right">-0.0560694</td>
</tr>
<tr class="odd">
<td align="left">EPHA1</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="linkages">Linkages<a class="anchor" aria-label="anchor" href="#linkages"></a>
</h3>
<p>Linkages between ligands, receptors, and transcription factors can be
accessed in several different ways, depending on the specific link and
the scope desired. The <code><a href="../reference/dom_linkages.html">dom_linkages()</a></code> function has three
arguments - the first, like all of our access functions, is for the
domino object. The second, <code>link_type</code>, is used to specify
which linkages are desired (options are complexes, receptor-ligand,
tf-target, or tf-receptor). The third argument, <code>by_cluster</code>,
determines whether the linkages returned are arranged by cluster (though
this does change the available linkage types to tf-receptor, receptor,
or incoming-ligand). For example, to access the complexes used across
the dataset:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">complex_links</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dom_linkages.html">dom_linkages</a></span><span class="op">(</span><span class="va">dom</span>, link_type <span class="op">=</span> <span class="st">"complexes"</span><span class="op">)</span></span>
<span><span class="co"># Look for components of NODAL receptor complex</span></span>
<span><span class="va">complex_links</span><span class="op">$</span><span class="va">NODAL_receptor</span></span>
<span><span class="co">#&gt; [1] "ACVR1B"       "ACVR2B"       "TDGF1"        "character(0)"</span></span></code></pre></div>
<p>To view incoming ligands to each cluster:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">incoming_links</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dom_linkages.html">dom_linkages</a></span><span class="op">(</span><span class="va">dom</span>, link_type <span class="op">=</span> <span class="st">"incoming-ligand"</span>, by_cluster <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co"># Check incoming signals to dendritic cells</span></span>
<span><span class="va">incoming_links</span><span class="op">$</span><span class="va">dendritic_cell</span></span>
<span><span class="co">#&gt;  [1] "MIF"          "APP"          "COPA"         "CD99"         "CD1D"        </span></span>
<span><span class="co">#&gt;  [6] "HLA-F"        "INS"          "ANXA1"        "CXCL9"        "BMP8B"       </span></span>
<span><span class="co">#&gt; [11] "a4b1_complex" "aVb3_complex" "TG"           "CSF3"         "IL34"        </span></span>
<span><span class="co">#&gt; [16] "CSF1"         "CSF2"         "TNF"          "GRN"          "LTA"         </span></span>
<span><span class="co">#&gt; [21] "CTLA4"        "CD28"</span></span></code></pre></div>
<p>If, for some reason, you find yourself in need of the entire linkage
structure (not recommended), it can be accessed through its slot name;
domino objects are S4 objects.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">all_linkages</span> <span class="op">&lt;-</span> <span class="fu">slot</span><span class="op">(</span><span class="va">dom</span>, <span class="st">"linkages"</span><span class="op">)</span></span>
<span><span class="co"># Names of all sub-structures:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">all_linkages</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "complexes"          "rec_lig"            "tf_targets"        </span></span>
<span><span class="co">#&gt; [4] "clust_tf"           "tf_rec"             "clust_tf_rec"      </span></span>
<span><span class="co">#&gt; [7] "clust_rec"          "clust_incoming_lig"</span></span></code></pre></div>
<p>Alternately, to obtain a simplified list of receptors, ligands,
and/or features in the domino object, use the
<code><a href="../reference/dom_network_items.html">dom_network_items()</a></code> function. To pull all transcription
factors associated with the dendritic cell cluster:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dc_tfs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dom_network_items.html">dom_network_items</a></span><span class="op">(</span><span class="va">dom</span>, <span class="st">"dendritic_cell"</span>, return <span class="op">=</span> <span class="st">"features"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">dc_tfs</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "ZNF189" "KLF12"  "ETV7"   "IRF8"   "KLF6"   "FOS"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="signaling-matrices">Signaling Matrices<a class="anchor" aria-label="anchor" href="#signaling-matrices"></a>
</h3>
<p>The averaged z-scored expression of ligands and receptors between
different clusters can be accessed in matrix form.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">signal_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dom_signaling.html">dom_signaling</a></span><span class="op">(</span><span class="va">dom</span><span class="op">)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">signal_matrix</span><span class="op">)</span></span></code></pre></div>
<table style="width:100%;" class="table">
<colgroup>
<col width="13%">
<col width="12%">
<col width="10%">
<col width="13%">
<col width="6%">
<col width="8%">
<col width="10%">
<col width="6%">
<col width="11%">
<col width="7%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="right">L_naive_CD4_T_cell</th>
<th align="right">L_CD14_monocyte</th>
<th align="right">L_memory_CD4_T_cell</th>
<th align="right">L_B_cell</th>
<th align="right">L_CD8_T_cell</th>
<th align="right">L_CD16_monocyte</th>
<th align="right">L_NK_cell</th>
<th align="right">L_dendritic_cell</th>
<th align="right">L_platelet</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">R_naive_CD4_T_cell</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.1711738</td>
<td align="right">0.1142025</td>
<td align="right">0.1316175</td>
<td align="right">0.2223212</td>
<td align="right">0.2429505</td>
<td align="right">0.000000</td>
<td align="right">0.00000</td>
</tr>
<tr class="even">
<td align="left">R_CD14_monocyte</td>
<td align="right">0.9054293</td>
<td align="right">2.2595922</td>
<td align="right">1.8484199</td>
<td align="right">0.4113900</td>
<td align="right">1.2793445</td>
<td align="right">1.6668345</td>
<td align="right">2.0653175</td>
<td align="right">4.629239</td>
<td align="right">3.38459</td>
</tr>
<tr class="odd">
<td align="left">R_memory_CD4_T_cell</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.1711738</td>
<td align="right">0.1142025</td>
<td align="right">0.1316175</td>
<td align="right">0.2223212</td>
<td align="right">0.2429505</td>
<td align="right">0.000000</td>
<td align="right">0.00000</td>
</tr>
<tr class="even">
<td align="left">R_B_cell</td>
<td align="right">0.0513875</td>
<td align="right">0.1920610</td>
<td align="right">0.2819420</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.2508179</td>
<td align="right">0.1639465</td>
<td align="right">1.052996</td>
<td align="right">2.47526</td>
</tr>
<tr class="odd">
<td align="left">R_CD8_T_cell</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.1711738</td>
<td align="right">0.1142025</td>
<td align="right">0.1316175</td>
<td align="right">0.2223212</td>
<td align="right">0.2429505</td>
<td align="right">0.000000</td>
<td align="right">0.00000</td>
</tr>
<tr class="even">
<td align="left">R_CD16_monocyte</td>
<td align="right">0.9054293</td>
<td align="right">2.2595922</td>
<td align="right">1.8484199</td>
<td align="right">0.4113900</td>
<td align="right">1.2793445</td>
<td align="right">1.6668345</td>
<td align="right">2.0653175</td>
<td align="right">4.629239</td>
<td align="right">3.38459</td>
</tr>
<tr class="odd">
<td align="left">R_NK_cell</td>
<td align="right">0.1127173</td>
<td align="right">1.5182660</td>
<td align="right">0.6536131</td>
<td align="right">0.0000000</td>
<td align="right">0.2272370</td>
<td align="right">1.0199925</td>
<td align="right">0.4168540</td>
<td align="right">2.582174</td>
<td align="right">2.47526</td>
</tr>
<tr class="even">
<td align="left">R_dendritic_cell</td>
<td align="right">0.2412594</td>
<td align="right">2.2595922</td>
<td align="right">1.4862873</td>
<td align="right">0.0583676</td>
<td align="right">1.2793445</td>
<td align="right">1.0699215</td>
<td align="right">2.0653175</td>
<td align="right">4.629239</td>
<td align="right">3.38459</td>
</tr>
<tr class="odd">
<td align="left">R_platelet</td>
<td align="right">0.4466970</td>
<td align="right">0.7381964</td>
<td align="right">0.5150958</td>
<td align="right">0.4113900</td>
<td align="right">0.3264903</td>
<td align="right">1.0468853</td>
<td align="right">0.5332812</td>
<td align="right">2.317655</td>
<td align="right">2.47526</td>
</tr>
</tbody>
</table>
<p>To view signaling to a specific cluster from the other clusters, set
the <code>cluster</code> argument to the cluster name.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dc_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dom_signaling.html">dom_signaling</a></span><span class="op">(</span><span class="va">dom</span>, <span class="st">"dendritic_cell"</span><span class="op">)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">dc_matrix</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="2%">
<col width="12%">
<col width="11%">
<col width="12%">
<col width="8%">
<col width="10%">
<col width="11%">
<col width="9%">
<col width="11%">
<col width="9%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="right">dendritic_cell.L_naive_CD4_T_cell</th>
<th align="right">dendritic_cell.L_CD14_monocyte</th>
<th align="right">dendritic_cell.L_memory_CD4_T_cell</th>
<th align="right">dendritic_cell.L_B_cell</th>
<th align="right">dendritic_cell.L_CD8_T_cell</th>
<th align="right">dendritic_cell.L_CD16_monocyte</th>
<th align="right">dendritic_cell.L_NK_cell</th>
<th align="right">dendritic_cell.L_dendritic_cell</th>
<th align="right">dendritic_cell.L_platelet</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">MIF</td>
<td align="right">0.0513875</td>
<td align="right">0.0000000</td>
<td align="right">0.2283356</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.1639465</td>
<td align="right">0.3383730</td>
<td align="right">0.0000000</td>
</tr>
<tr class="even">
<td align="left">APP</td>
<td align="right">0.0000000</td>
<td align="right">0.0910785</td>
<td align="right">0.0478541</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.5651886</td>
<td align="right">2.0283841</td>
</tr>
<tr class="odd">
<td align="left">COPA</td>
<td align="right">0.0000000</td>
<td align="right">0.1009825</td>
<td align="right">0.0057523</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.2508179</td>
<td align="right">0.0000000</td>
<td align="right">0.1494349</td>
<td align="right">0.4468762</td>
</tr>
<tr class="even">
<td align="left">CD99</td>
<td align="right">0.0000000</td>
<td align="right">0.0249929</td>
<td align="right">0.1941752</td>
<td align="right">0.0000000</td>
<td align="right">0.5222139</td>
<td align="right">0.0000000</td>
<td align="right">0.9530171</td>
<td align="right">0.1656705</td>
<td align="right">0.9093296</td>
</tr>
<tr class="odd">
<td align="left">CD1D</td>
<td align="right">0.0000000</td>
<td align="right">0.5461355</td>
<td align="right">0.0000000</td>
<td align="right">0.0583676</td>
<td align="right">0.0000000</td>
<td align="right">0.1733057</td>
<td align="right">0.0000000</td>
<td align="right">1.2646585</td>
<td align="right">0.0000000</td>
</tr>
<tr class="even">
<td align="left">HLA-F</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0133417</td>
<td align="right">0.0000000</td>
<td align="right">0.3264903</td>
<td align="right">0.0258488</td>
<td align="right">0.3693347</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
</tr>
<tr class="odd">
<td align="left">ANXA1</td>
<td align="right">0.0000000</td>
<td align="right">0.1618442</td>
<td align="right">0.4100748</td>
<td align="right">0.0000000</td>
<td align="right">0.2618538</td>
<td align="right">0.0687405</td>
<td align="right">0.5600050</td>
<td align="right">0.6167362</td>
<td align="right">0.0000000</td>
</tr>
<tr class="even">
<td align="left">CXCL9</td>
<td align="right">0.0000000</td>
<td align="right">0.0083536</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
</tr>
<tr class="odd">
<td align="left">CSF1</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0487887</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0043552</td>
<td align="right">0.0090572</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
</tr>
<tr class="even">
<td align="left">TNF</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0543116</td>
<td align="right">0.0000000</td>
<td align="right">0.1574295</td>
<td align="right">0.0589903</td>
<td align="right">0.0000000</td>
<td align="right">0.0061102</td>
<td align="right">0.0000000</td>
</tr>
<tr class="odd">
<td align="left">GRN</td>
<td align="right">0.0000000</td>
<td align="right">1.3262050</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.4878632</td>
<td align="right">0.0000000</td>
<td align="right">1.5230676</td>
<td align="right">0.0000000</td>
</tr>
<tr class="even">
<td align="left">LTA</td>
<td align="right">0.0613298</td>
<td align="right">0.0000000</td>
<td align="right">0.1746769</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0099570</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
</tr>
<tr class="odd">
<td align="left">CTLA4</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0638344</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
</tr>
<tr class="even">
<td align="left">CD28</td>
<td align="right">0.1285421</td>
<td align="right">0.0000000</td>
<td align="right">0.2451419</td>
<td align="right">0.0000000</td>
<td align="right">0.0113571</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="build-information">Build information<a class="anchor" aria-label="anchor" href="#build-information"></a>
</h3>
<p>To keep track of the options set when running
<code><a href="../reference/build_domino.html">build_domino()</a></code>, they are stored within the domino object
itself. To view these options, use the <code><a href="../reference/dom_info.html">dom_info()</a></code>
function.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/dom_info.html">dom_info</a></span><span class="op">(</span><span class="va">dom</span><span class="op">)</span></span>
<span><span class="co">#&gt; $create</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $build</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $build_variables</span></span>
<span><span class="co">#&gt;     max_tf_per_clust          min_tf_pval       max_rec_per_tf </span></span>
<span><span class="co">#&gt;               25.000                0.001               25.000 </span></span>
<span><span class="co">#&gt; rec_tf_cor_threshold   min_rec_percentage </span></span>
<span><span class="co">#&gt;                0.250                0.100</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="continued-development">Continued Development<a class="anchor" aria-label="anchor" href="#continued-development"></a>
</h2>
<p>Since domino2 is a package still being developed, there are new
functions and features that will be implemented in future versions. In
the meantime, we have put together further information on <a href="vignette(%22plotting_vignette%22)">plotting</a> and an example
analysis can be viewed on our <a href="vignette(%22domino2%22)">Getting
Started</a> page. Additionally, if you find any bugs, have further
questions, or want to share an idea, please let us know <a href="https://github.com/FertigLab/domino_development/issues" class="external-link">here</a>.</p>
<details><summary>
Vignette Build Information
</summary><p>Date last built and session information:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.Date</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "2024-01-05"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.2 (2023-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] domino2_0.99.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] DBI_1.2.0               bitops_1.0-7            formatR_1.14           </span></span>
<span><span class="co">#&gt;   [4] biomaRt_2.58.0          rlang_1.1.2             magrittr_2.0.3         </span></span>
<span><span class="co">#&gt;   [7] clue_0.3-65             GetoptLong_1.0.5        matrixStats_1.2.0      </span></span>
<span><span class="co">#&gt;  [10] compiler_4.3.2          RSQLite_2.3.4           png_0.1-8              </span></span>
<span><span class="co">#&gt;  [13] systemfonts_1.0.5       vctrs_0.6.5             stringr_1.5.1          </span></span>
<span><span class="co">#&gt;  [16] pkgconfig_2.0.3         shape_1.4.6             crayon_1.5.2           </span></span>
<span><span class="co">#&gt;  [19] fastmap_1.1.1           backports_1.4.1         dbplyr_2.4.0           </span></span>
<span><span class="co">#&gt;  [22] XVector_0.42.0          utf8_1.2.4              rmarkdown_2.25         </span></span>
<span><span class="co">#&gt;  [25] ragg_1.2.7              purrr_1.0.2             bit_4.0.5              </span></span>
<span><span class="co">#&gt;  [28] xfun_0.41               zlibbioc_1.48.0         cachem_1.0.8           </span></span>
<span><span class="co">#&gt;  [31] GenomeInfoDb_1.38.5     jsonlite_1.8.8          progress_1.2.3         </span></span>
<span><span class="co">#&gt;  [34] blob_1.2.4              broom_1.0.5             parallel_4.3.2         </span></span>
<span><span class="co">#&gt;  [37] prettyunits_1.2.0       cluster_2.1.4           R6_2.5.1               </span></span>
<span><span class="co">#&gt;  [40] bslib_0.6.1             stringi_1.8.3           RColorBrewer_1.1-3     </span></span>
<span><span class="co">#&gt;  [43] car_3.1-2               jquerylib_0.1.4         Rcpp_1.0.11            </span></span>
<span><span class="co">#&gt;  [46] iterators_1.0.14        knitr_1.45              IRanges_2.36.0         </span></span>
<span><span class="co">#&gt;  [49] igraph_1.6.0            Matrix_1.6-4            tidyselect_1.2.0       </span></span>
<span><span class="co">#&gt;  [52] abind_1.4-5             yaml_2.3.8              doParallel_1.0.17      </span></span>
<span><span class="co">#&gt;  [55] codetools_0.2-19        curl_5.2.0              lattice_0.21-9         </span></span>
<span><span class="co">#&gt;  [58] tibble_3.2.1            plyr_1.8.9              Biobase_2.62.0         </span></span>
<span><span class="co">#&gt;  [61] KEGGREST_1.42.0         evaluate_0.23           desc_1.4.3             </span></span>
<span><span class="co">#&gt;  [64] BiocFileCache_2.10.1    xml2_1.3.6              circlize_0.4.15        </span></span>
<span><span class="co">#&gt;  [67] Biostrings_2.70.1       pillar_1.9.0            ggpubr_0.6.0           </span></span>
<span><span class="co">#&gt;  [70] filelock_1.0.3          carData_3.0-5           foreach_1.5.2          </span></span>
<span><span class="co">#&gt;  [73] stats4_4.3.2            generics_0.1.3          RCurl_1.98-1.13        </span></span>
<span><span class="co">#&gt;  [76] S4Vectors_0.40.2        hms_1.1.3               ggplot2_3.4.4          </span></span>
<span><span class="co">#&gt;  [79] munsell_0.5.0           scales_1.3.0            glue_1.6.2             </span></span>
<span><span class="co">#&gt;  [82] tools_4.3.2             ggsignif_0.6.4          fs_1.6.3               </span></span>
<span><span class="co">#&gt;  [85] XML_3.99-0.16           grid_4.3.2              tidyr_1.3.0            </span></span>
<span><span class="co">#&gt;  [88] AnnotationDbi_1.64.1    colorspace_2.1-0        GenomeInfoDbData_1.2.11</span></span>
<span><span class="co">#&gt;  [91] cli_3.6.2               rappdirs_0.3.3          textshaping_0.3.7      </span></span>
<span><span class="co">#&gt;  [94] fansi_1.0.6             ComplexHeatmap_2.18.0   dplyr_1.1.4            </span></span>
<span><span class="co">#&gt;  [97] gtable_0.3.4            rstatix_0.7.2           sass_0.4.8             </span></span>
<span><span class="co">#&gt; [100] digest_0.6.33           BiocGenerics_0.48.1     rjson_0.2.21           </span></span>
<span><span class="co">#&gt; [103] memoise_2.0.1           htmltools_0.5.7         pkgdown_2.0.7          </span></span>
<span><span class="co">#&gt; [106] lifecycle_1.0.4         httr_1.4.7              GlobalOptions_0.1.2    </span></span>
<span><span class="co">#&gt; [109] bit64_4.0.5</span></span></code></pre></div>
</details>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>domino2 is developed by Christopher Cherry, Jacob T Mitchell, Sushma Nagaraj, Kavita Krishnan, Dmitrijs Lvovs.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
